// terra-application/lib/disclosure-manager/index.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DisclosureManagerContext", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.DisclosureManagerContext;
  }
});
Object.defineProperty(exports, "DisclosureManagerDelegate", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.DisclosureManagerDelegate;
  }
});
Object.defineProperty(exports, "DisclosureManagerHeaderAdapter", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.DisclosureManagerHeaderAdapter;
  }
});
Object.defineProperty(exports, "DisclosureManagerHeaderAdapterContext", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.DisclosureManagerHeaderAdapterContext;
  }
});
Object.defineProperty(exports, "availableDisclosureHeights", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.availableDisclosureHeights;
  }
});
Object.defineProperty(exports, "availableDisclosureSizes", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.availableDisclosureSizes;
  }
});
Object.defineProperty(exports, "availableDisclosureWidths", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.availableDisclosureWidths;
  }
});
Object.defineProperty(exports, "closeMostRecentDisclosure", {
  enumerable: true,
  get: function get() {
    return _disclosureCallbacks.closeMostRecentDisclosure;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.default;
  }
});
Object.defineProperty(exports, "disclosureManagerShape", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.disclosureManagerShape;
  }
});
Object.defineProperty(exports, "getActiveDisclosureCount", {
  enumerable: true,
  get: function get() {
    return _disclosureCallbacks.getActiveDisclosureCount;
  }
});
Object.defineProperty(exports, "withDisclosureManager", {
  enumerable: true,
  get: function get() {
    return _terraDisclosureManager.withDisclosureManager;
  }
});
var _terraDisclosureManager = _interopRequireWildcard(require("terra-disclosure-manager"));
var _disclosureCallbacks = require("./_disclosureCallbacks");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
