// terra-application/lib/workspace/subcomponents/_Tab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _keycodeJs = require("keycode-js");
var _TabUtils = require("./_TabUtils");
var _TabModule = _interopRequireDefault(require("./Tab.module.scss"));
var cx = _bind.default.bind(_TabModule.default);
var propTypes = {
  /**
   * The id of the tab to be used in mapping.
   */
  id: _propTypes.default.string.isRequired,
  /**
   * The id of the tab pane element associated to this tab.
   */
  associatedPanelId: _propTypes.default.string.isRequired,
  /**
   * Index value to use for navigation.
   */
  index: _propTypes.default.number.isRequired,
  /**
   * Indicates if the tab is currently selected.
   */
  isSelected: _propTypes.default.bool,
  /**
   * Text to be displayed on the tab.
   */
  label: _propTypes.default.string.isRequired,
  /**
   * Callback function triggering on selection. onSelect(itemKey, metaData)
   */
  onSelect: _propTypes.default.func.isRequired,
  /**
   * Identifier for the workspace item represented by the Tab. Returned with onSelect.
   */
  itemKey: _propTypes.default.string,
  /**
   * Data for the workspace item represented by the Tab. Returned with onSelect.
   */
  metaData: _propTypes.default.object,
  /**
   * Array of id strings,
   */
  tabIds: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,
  /**
   * The z-index style to apply to the tab based upon order and state.
   */
  zIndex: _propTypes.default.number
};
var defaultProps = {
  isSelected: false
};
var Tab = function Tab(_ref) {
  var id = _ref.id,
    associatedPanelId = _ref.associatedPanelId,
    itemKey = _ref.itemKey,
    index = _ref.index,
    isSelected = _ref.isSelected,
    label = _ref.label,
    metaData = _ref.metaData,
    onSelect = _ref.onSelect,
    tabIds = _ref.tabIds,
    zIndex = _ref.zIndex;
  var attributes = {};
  var theme = _react.default.useContext(_terraThemeContext.default);
  var tabClassNames = cx('tab', {
    'is-active': isSelected
  }, theme.className);
  function onKeyDown(event) {
    if (event.nativeEvent.keyCode === _keycodeJs.KEY_RETURN || event.nativeEvent.keyCode === _keycodeJs.KEY_SPACE) {
      event.preventDefault();
      event.stopPropagation();
      onSelect(itemKey, metaData);
    } else {
      (0, _TabUtils.handleArrows)(event, index, tabIds);
    }
  }
  function onClick() {
    onSelect(itemKey, metaData);
  }
  attributes.tabIndex = isSelected ? 0 : -1;
  attributes.onClick = onClick;
  attributes.onKeyDown = onKeyDown;
  attributes.onBlur = _TabUtils.enableFocusStyles;
  attributes.onMouseDown = _TabUtils.disableFocusStyles;
  attributes['data-focus-styles-enabled'] = true;
  attributes['aria-selected'] = isSelected;
  attributes.style = {
    zIndex: zIndex
  };
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, attributes, {
    id: id,
    "aria-controls": associatedPanelId,
    role: "tab",
    className: tabClassNames,
    title: label
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: cx('inner')
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cx('label')
  }, label)));
};
Tab.propTypes = propTypes;
Tab.defaultProps = defaultProps;
var _default = Tab;
exports.default = _default;
