// terra-application/lib/terra-dev-site/app/components.2/example/ApplicationErrorBoundaryExample.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _ApplicationErrorBoundary = _interopRequireDefault(require("terra-application/lib/application-error-boundary/ApplicationErrorBoundary"));
var _ApplicationErrorBoundaryExampleModule = _interopRequireDefault(require("./ApplicationErrorBoundaryExample.module.scss"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var cx = _bind.default.bind(_ApplicationErrorBoundaryExampleModule.default);
var ErrorThrowingComponent = function ErrorThrowingComponent() {
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    throwError = _useState2[0],
    setThrowError = _useState2[1];
  if (throwError) {
    throw new Error('whoops');
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, "Press the button below to trigger an error."), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      setThrowError(true);
    }
  }, "Trigger Error"));
};
var ApplicationErrorBoundaryExampleWrapper = function ApplicationErrorBoundaryExampleWrapper(_ref) {
  var children = _ref.children;
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    refresh = _useState4[0],
    setRefresh = _useState4[1];
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('example')
  }, /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      setRefresh(!refresh);
    }
  }, "Refresh Component"), /*#__PURE__*/_react.default.createElement(_ApplicationErrorBoundary.default, null, children));
};
ApplicationErrorBoundaryExampleWrapper.propTypes = {
  children: _propTypes.default.node
};
var ApplicationErrorBoundaryExample = function ApplicationErrorBoundaryExample() {
  return /*#__PURE__*/_react.default.createElement(ApplicationErrorBoundaryExampleWrapper, null, /*#__PURE__*/_react.default.createElement(ErrorThrowingComponent, null));
};
var _default = ApplicationErrorBoundaryExample;
exports.default = _default;
