// terra-application/lib/terra-dev-site/app/components.2/example/ApplicationNavigationExample.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _applicationNavigation = _interopRequireDefault(require("terra-application/lib/application-navigation"));
var _navigationPrompt = _interopRequireDefault(require("terra-application/lib/navigation-prompt"));
var _ApplicationNavigationExampleModule = _interopRequireDefault(require("./ApplicationNavigationExample.module.scss"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var cx = _bind.default.bind(_ApplicationNavigationExampleModule.default);
var ExamplePageContent = function ExamplePageContent(_ref) {
  var title = _ref.title;
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    hasPendingAction = _useState2[0],
    setHasPendingAction = _useState2[1];
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('page-content')
  }, /*#__PURE__*/_react.default.createElement("p", null, title), /*#__PURE__*/_react.default.createElement("p", null, "Toggle pending action:", ' ', /*#__PURE__*/_react.default.createElement("button", {
    id: "pending-action-toggle",
    type: "button",
    onClick: function onClick() {
      setHasPendingAction(!hasPendingAction);
    }
  }, hasPendingAction ? 'Disable' : 'Enable')), hasPendingAction ? /*#__PURE__*/_react.default.createElement(_navigationPrompt.default, {
    description: "Testing ApplicationNavigation's navigation prompt handling"
  }) : undefined);
};
ExamplePageContent.propTypes = {
  title: _propTypes.default.string
};
var navigationItems = [{
  key: 'page_1',
  text: 'Page 1'
}, {
  key: 'page_2',
  text: 'Page 2'
}, {
  key: 'page_3',
  text: 'Page 3'
}];
var userConfig = {
  name: 'Example User',
  initials: 'EU',
  detail: 'Is a User'
};
var userActionConfig = {
  text: 'Edit Photo',
  // eslint-disable-next-line no-alert
  userActionCallback: function userActionCallback() {
    return alert('Edit Photo Clicked');
  }
};
var ApplicationNavigationTest = function ApplicationNavigationTest() {
  var _useState3 = (0, _react.useState)('page_1'),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    activeNavItem = _useState4[0],
    setActiveNavItem = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    loggedOut = _useState6[0],
    setLoggedOut = _useState6[1];
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('example-container')
  }, loggedOut ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, "Logged Out"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      setLoggedOut(false);
    }
  }, "Reset")) : /*#__PURE__*/_react.default.createElement(_applicationNavigation.default, {
    titleConfig: {
      title: 'ApplicationNavigation Example'
    },
    userConfig: userConfig,
    userActionConfig: userActionConfig,
    navigationItems: navigationItems,
    activeNavigationItemKey: activeNavItem,
    onSelectNavigationItem: function onSelectNavigationItem(key) {
      setActiveNavItem(key);
    },
    onSelectLogout: function onSelectLogout() {
      setLoggedOut(true);
    }
  }, /*#__PURE__*/_react.default.createElement(ExamplePageContent, {
    key: activeNavItem,
    title: activeNavItem
  })));
};
var _default = ApplicationNavigationTest;
exports.default = _default;
