// terra-application/lib/terra-dev-site/app/components.2/example/ApplicationBaseExample.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _applicationIntl = require("terra-application/lib/application-intl");
var _theme = require("terra-application/lib/theme");
var _breakpoints = require("terra-application/lib/breakpoints");
var _applicationLoadingOverlay = _interopRequireDefault(require("terra-application/lib/application-loading-overlay"));
var _applicationBase = _interopRequireDefault(require("terra-application/lib/application-base"));
var _navigationPrompt = _interopRequireDefault(require("terra-application/lib/navigation-prompt"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var ApplicationContentExample = function ApplicationContentExample() {
  var _useState = (0, _react.useState)(false),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    isLoading = _useState2[0],
    setIsLoading = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    throwError = _useState4[0],
    setThrowError = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    blockUnload = _useState6[0],
    setBlockUnload = _useState6[1];
  var activeBreakpoint = (0, _react.useContext)(_breakpoints.ActiveBreakpointContext);
  var applicationIntl = (0, _react.useContext)(_applicationIntl.ApplicationIntlContext);
  if (throwError) {
    throw new Error("Testing ApplicationBase's error boundary...");
  }
  var activeBreakpointExample = /*#__PURE__*/_react.default.createElement("p", null, "Active Breakpoint:", ' ', activeBreakpoint);
  var intlExample = /*#__PURE__*/_react.default.createElement("p", null, "Active Locale:", ' ', applicationIntl.locale);
  var errorBoundaryExample = /*#__PURE__*/_react.default.createElement("p", null, "Render Error:", ' ', /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      setThrowError(true);
    }
  }, "Throw"));
  var loadingOverlayExample = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, "Show Loading Overlay (3 seconds):", ' ', /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      setIsLoading(true);
      setTimeout(function () {
        setIsLoading(false);
      }, 3000);
    }
  }, "Show")), /*#__PURE__*/_react.default.createElement(_applicationLoadingOverlay.default, {
    isOpen: isLoading,
    message: "Testing ApplicationBase's loading overlay",
    backgroundStyle: "light"
  }));
  var navigationPromptExample = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, "Prompt on window close:", ' ', /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    name: "prompt",
    onChange: function onChange() {
      setBlockUnload(!blockUnload);
    }
  })), blockUnload ? /*#__PURE__*/_react.default.createElement(_navigationPrompt.default, {
    description: "Testing ApplicationBase's navigation prompt handling"
  }) : undefined);
  return /*#__PURE__*/_react.default.createElement("div", null, activeBreakpointExample, intlExample, errorBoundaryExample, loadingOverlayExample, navigationPromptExample);
};
var ApplicationBaseExample = function ApplicationBaseExample() {
  var applicationIntl = (0, _react.useContext)(_applicationIntl.ApplicationIntlContext);
  var theme = _react.default.useContext(_theme.ThemeContext);
  return /*#__PURE__*/_react.default.createElement(_applicationBase.default, {
    locale: applicationIntl.locale,
    themeName: theme.className
  }, /*#__PURE__*/_react.default.createElement(ApplicationContentExample, null));
};
var _default = ApplicationBaseExample;
exports.default = _default;
