// terra-breakpoints/lib/breakpoints.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.breakpointIsActiveForSize = exports.activeBreakpointForSize = void 0;
var breakpoints = {
  tiny: 0,
  small: 544,
  medium: 768,
  large: 992,
  huge: 1216,
  enormous: 1440
};
var breakpointSequence = ['tiny', 'small', 'medium', 'large', 'huge', 'enormous'];
var activeBreakpointForSize = function activeBreakpointForSize(size) {
  for (var i = breakpointSequence.length - 1; i >= 0; i -= 1) {
    var breakpoint = breakpointSequence[i];
    if (size >= breakpoints[breakpoint]) {
      return breakpoint;
    }
  }
  return undefined;
};
exports.activeBreakpointForSize = activeBreakpointForSize;
var breakpointIsActiveForSize = function breakpointIsActiveForSize(breakpoint, size) {
  var breakpointLocation = breakpointSequence.indexOf(breakpoint);
  var sizeLocation = breakpointSequence.indexOf(activeBreakpointForSize(size));
  return sizeLocation === -1 || breakpointLocation === -1 ? false : sizeLocation >= breakpointLocation;
};
exports.breakpointIsActiveForSize = breakpointIsActiveForSize;
var _default = breakpoints;
exports.default = _default;
