// @cerner/terra-dev-site/lib/webpack/plugin/applyDefaults.js
"use strict";

var startCase = require('lodash.startcase');
var fs = require('fs');
var path = require('path');
var npmPackage = JSON.parse(fs.readFileSync(path.join(process.cwd(), 'package.json')));
var applyDefaults = function applyDefaults() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref$primaryNavigatio = _ref.primaryNavigationItems,
    primaryNavigationItems = _ref$primaryNavigatio === void 0 ? [{
      path: '/home',
      label: 'Home',
      contentExtension: 'home',
      additionalContent: [{
        label: 'Home',
        filePath: path.resolve(process.cwd(), 'README.md')
      }]
    }, {
      path: '/application',
      label: 'Application',
      contentExtension: 'app'
    }, {
      path: '/components',
      label: 'Components',
      contentExtension: 'doc'
    }, {
      path: '/dev_tools',
      label: 'Developer Tools',
      contentExtension: 'tool'
    }, {
      path: '/guides',
      label: 'Guides',
      contentExtension: 'guide'
    }, {
      path: '/tests',
      label: 'Tests',
      contentExtension: 'test'
    }] : _ref$primaryNavigatio,
    _ref$additionalSearch = _ref.additionalSearchDirectories,
    additionalSearchDirectories = _ref$additionalSearch === void 0 ? [] : _ref$additionalSearch,
    _ref$sideEffectImport = _ref.sideEffectImportFilePaths,
    sideEffectImportFilePaths = _ref$sideEffectImport === void 0 ? [] : _ref$sideEffectImport,
    _ref$polyFillSideEffe = _ref.polyFillSideEffectImportFilePath,
    polyFillSideEffectImportFilePath = _ref$polyFillSideEffe === void 0 ? '@cerner/terra-polyfill' : _ref$polyFillSideEffe,
    _ref$titleConfig = _ref.titleConfig,
    titleConfig = _ref$titleConfig === void 0 ? {
      title: startCase(npmPackage.name.split('/').pop())
    } : _ref$titleConfig,
    defaultLocale = _ref.defaultLocale,
    defaultTheme = _ref.defaultTheme,
    defaultDirection = _ref.defaultDirection,
    _ref$faviconFilePath = _ref.faviconFilePath,
    faviconFilePath = _ref$faviconFilePath === void 0 ? path.join(__dirname, '..', 'terra-favicon', '32px', 'favicon.ico') : _ref$faviconFilePath,
    _ref$extensionItems = _ref.extensionItems,
    extensionItems = _ref$extensionItems === void 0 ? [] : _ref$extensionItems,
    _ref$headHtml = _ref.headHtml,
    headHtml = _ref$headHtml === void 0 ? [] : _ref$headHtml,
    pathPrefix = _ref.pathPrefix,
    _ref$sourceFolder = _ref.sourceFolder,
    sourceFolder = _ref$sourceFolder === void 0 ? 'src' : _ref$sourceFolder,
    _ref$distributionFold = _ref.distributionFolder,
    distributionFolder = _ref$distributionFold === void 0 ? 'lib' : _ref$distributionFold,
    _ref$enableDebugLoggi = _ref.enableDebugLogging,
    enableDebugLogging = _ref$enableDebugLoggi === void 0 ? false : _ref$enableDebugLoggi,
    _ref$additionalConfig = _ref.additionalConfig,
    additionalConfig = _ref$additionalConfig === void 0 ? {} : _ref$additionalConfig,
    _ref$excludeChunks = _ref.excludeChunks,
    excludeChunks = _ref$excludeChunks === void 0 ? [] : _ref$excludeChunks,
    _ref$contentDirectory = _ref.contentDirectory,
    contentDirectory = _ref$contentDirectory === void 0 ? 'terra-dev-site' : _ref$contentDirectory;
  // Make the polyfill path the first side effect if present.
  var sideEffects = sideEffectImportFilePaths;
  if (polyFillSideEffectImportFilePath !== 'none') {
    sideEffects.unshift(polyFillSideEffectImportFilePath);
  }
  return {
    primaryNavigationItems: primaryNavigationItems,
    additionalSearchDirectories: additionalSearchDirectories,
    sideEffectImportFilePaths: sideEffects,
    titleConfig: titleConfig,
    defaultLocale: defaultLocale,
    defaultTheme: defaultTheme,
    defaultDirection: defaultDirection,
    faviconFilePath: faviconFilePath,
    extensionItems: extensionItems,
    headHtml: headHtml,
    pathPrefix: pathPrefix,
    sourceFolder: sourceFolder,
    distributionFolder: distributionFolder,
    enableDebugLogging: enableDebugLogging,
    namespace: npmPackage.name,
    additionalConfig: additionalConfig,
    excludeChunks: excludeChunks,
    contentDirectory: contentDirectory
  };
};
module.exports = applyDefaults;
