// terra-application/lib/action-menu/_ActionMenuHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _reactIntl = require("react-intl");
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _terraButton = _interopRequireDefault(require("terra-button"));
var _IconClose = _interopRequireDefault(require("terra-icon/lib/icon/IconClose"));
var _ActionMenuHeaderModule = _interopRequireDefault(require("./ActionMenuHeader.module.scss"));
var cx = _bind.default.bind(_ActionMenuHeaderModule.default);
var propTypes = {
  /**
   * The label text to display for the header.
   */
  label: _propTypes.default.string.isRequired,
  /**
   * Callback function for close button selection.
   * Returns the event e.g. onClose(event).
   */
  onClose: _propTypes.default.func,
  /**
   * @private
   */
  intl: _reactIntl.intlShape
};
var ActionMenuHeader = function ActionMenuHeader(_ref) {
  var label = _ref.label,
    onClose = _ref.onClose,
    intl = _ref.intl;
  var theme = _react.default.useContext(_terraThemeContext.default);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('action-menu-header', theme.className)
  }, /*#__PURE__*/_react.default.createElement("div", {
    "aria-hidden": true,
    className: cx('title')
  }, label), /*#__PURE__*/_react.default.createElement("div", {
    className: cx('close-button')
  }, onClose ? /*#__PURE__*/_react.default.createElement(_terraButton.default, {
    icon: /*#__PURE__*/_react.default.createElement(_IconClose.default, null),
    text: intl.formatMessage({
      id: 'terraApplication.actionMenu.headerCloseButtonLabel'
    }),
    onClick: onClose,
    variant: "utility"
  }) : undefined));
};
ActionMenuHeader.propTypes = propTypes;
var _default = (0, _reactIntl.injectIntl)(ActionMenuHeader);
exports.default = _default;
