// terra-application/lib/action-menu/ActionMenuRadio.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _IconCheckmark = _interopRequireDefault(require("terra-icon/lib/icon/IconCheckmark"));
var _ActionUtils = require("./_ActionUtils");
var _ActionMenuRadioModule = _interopRequireDefault(require("./ActionMenuRadio.module.scss"));
var cx = _bind.default.bind(_ActionMenuRadioModule.default);
var propTypes = {
  /**
   * The string used as an identifier for keyboard navigation.
   */
  actionKey: _propTypes.default.string.isRequired,
  /**
   * Optional icon to place with the radio.
   */
  icon: _propTypes.default.element,
  /**
   * Whether or not the radio is disabled.
   */
  isDisabled: _propTypes.default.bool,
  /**
   * Whether or not the radio is checked.
   */
  isChecked: _propTypes.default.bool,
  /**
   * The label text to display for the radio.
   */
  label: _propTypes.default.string.isRequired,
  /**
   * Callback function for action element selection.
   * Returns the event e.g. onAction(event).
   */
  onAction: _propTypes.default.func,
  /**
   * @private
   * Callback function for arrow key interaction.
   * Returns the event e.g. onArrow(key, direction).
   */
  onArrow: _propTypes.default.func,
  /**
   * @private
   * Callback function for arrow char.
   * Returns the event e.g. onChar(key, char).
   */
  onChar: _propTypes.default.func
};
var ActionMenuRadio = function ActionMenuRadio(_ref) {
  var actionKey = _ref.actionKey,
    icon = _ref.icon,
    isDisabled = _ref.isDisabled,
    isChecked = _ref.isChecked,
    label = _ref.label,
    onAction = _ref.onAction,
    onArrow = _ref.onArrow,
    onChar = _ref.onChar;
  var attrs = {};
  if (isDisabled) {
    attrs['aria-disabled'] = true;
  } else {
    attrs.tabIndex = '-1';
    attrs.onClick = onAction;
    attrs.onKeyDown = (0, _ActionUtils.generateOnKeyDown)(actionKey, onAction, onArrow, onChar);
    attrs.onBlur = _ActionUtils.enableFocusStyles;
    attrs.onMouseDown = _ActionUtils.disableFocusStyles;
    attrs['data-focus-styles-enabled'] = true;
  }
  var theme = _react.default.useContext(_terraThemeContext.default);
  /* eslint-disable jsx-a11y/no-noninteractive-element-to-interactive-role */
  return /*#__PURE__*/_react.default.createElement("li", (0, _extends2.default)({}, attrs, {
    className: cx('action-menu-radio', {
      'is-checked': isChecked
    }, {
      'is-actionable': !isDisabled
    }, {
      'is-disabled': isDisabled
    }, theme.className),
    role: "menuitemradio",
    "aria-checked": isChecked,
    "data-action-menu-key": actionKey
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: cx('checkbox')
  }, isChecked ? /*#__PURE__*/_react.default.createElement(_IconCheckmark.default, null) : null), icon ? /*#__PURE__*/_react.default.createElement("div", {
    className: cx('icon')
  }, icon) : null, /*#__PURE__*/_react.default.createElement("div", {
    className: cx('content')
  }, label));
  /* eslint-enable jsx-a11y/no-noninteractive-element-to-interactive-role */
};

ActionMenuRadio.propTypes = propTypes;
ActionMenuRadio.interactiveType = true;
var _default = ActionMenuRadio;
exports.default = _default;
