// terra-application/lib/application-status-overlay/ApplicationStatusOverlayProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraScroll = _interopRequireDefault(require("terra-scroll"));
var _terraStatusView = _interopRequireDefault(require("terra-status-view"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _ApplicationStatusOverlayContext = _interopRequireDefault(require("./ApplicationStatusOverlayContext"));
var _ApplicationStatusOverlayProviderModule = _interopRequireDefault(require("./ApplicationStatusOverlayProvider.module.scss"));
var _excluded = ["children", "scrollRefCallback"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var cx = _bind.default.bind(_ApplicationStatusOverlayProviderModule.default);
var propTypes = {
  /**
   * The components to be rendered within the context of the ApplicationStatusOverlayProvider.
   * Components rendered here are able to interact with ApplicationStatusOverlayProvider through
   * the ApplicationStatusOverlayContext.
   */
  children: _propTypes.default.node,
  /**
   * A function to be called with the current ref of the scrollable element rendered within the
   * ApplicationStatusOverlayProvider.
   */
  scrollRefCallback: _propTypes.default.func
};
var ApplicationStatusOverlayProvider = function ApplicationStatusOverlayProvider(_ref) {
  var _registeredStatusOver, _registeredStatusOver2, _registeredStatusOver3;
  var children = _ref.children,
    scrollRefCallback = _ref.scrollRefCallback,
    customProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var _React$useState = _react.default.useState({}),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    registeredStatusOverlay = _React$useState2[0],
    setRegisteredStatusOverlay = _React$useState2[1];
  var containerRef = _react.default.useRef();
  var contextValue = (0, _react.useMemo)(function () {
    return {
      show: function show(key, data) {
        setRegisteredStatusOverlay(function (state) {
          return _objectSpread(_objectSpread({}, state), {}, (0, _defineProperty2.default)({}, "".concat(key), data));
        });
      },
      hide: function hide(key) {
        setRegisteredStatusOverlay(function (state) {
          var newRegisteredStatusOverlay = _objectSpread({}, state);
          delete newRegisteredStatusOverlay[key];
          return newRegisteredStatusOverlay;
        });
      }
    };
  }, []);
  var disableContainerChildrenFocus = function disableContainerChildrenFocus() {
    if (containerRef.current) {
      containerRef.current.setAttribute('inert', '');
    }
  };
  var enableContainerChildrenFocus = function enableContainerChildrenFocus() {
    if (containerRef.current) {
      containerRef.current.removeAttribute('inert');
      containerRef.current.removeAttribute('aria-hidden');
    }
  };
  var registeredStatusOverlayKeys = Object.keys(registeredStatusOverlay);
  // In the event when multiple ApplicationStatusOverlay's are provided, the last rendered wins
  var lastRegisteredStatusOverlayKey = registeredStatusOverlayKeys.length && registeredStatusOverlayKeys[registeredStatusOverlayKeys.length - 1];
  var registeredButtonAttrs = (_registeredStatusOver = registeredStatusOverlay[lastRegisteredStatusOverlayKey]) === null || _registeredStatusOver === void 0 ? void 0 : _registeredStatusOver.buttonAttrs;
  var registeredMessage = (_registeredStatusOver2 = registeredStatusOverlay[lastRegisteredStatusOverlayKey]) === null || _registeredStatusOver2 === void 0 ? void 0 : _registeredStatusOver2.message;
  var registeredVariant = (_registeredStatusOver3 = registeredStatusOverlay[lastRegisteredStatusOverlayKey]) === null || _registeredStatusOver3 === void 0 ? void 0 : _registeredStatusOver3.variant;
  (0, _react.useEffect)(function () {
    // eslint-disable-next-line no-prototype-builtins
    if (!Element.prototype.hasOwnProperty('inert')) {
      // IE10 throws an error if wicg-inert is imported too early, as wicg-inert tries to set an observer on document.body which may not exist on import
      // eslint-disable-next-line global-require
      require('wicg-inert/dist/inert');
    }
  });
  (0, _react.useEffect)(function () {
    if (registeredStatusOverlayKeys.length) {
      disableContainerChildrenFocus();
    } else {
      enableContainerChildrenFocus();
    }
  }, [registeredStatusOverlayKeys]);
  var className = cx('container');
  if (customProps.className) {
    className = [className, customProps.className].join(' ');
  }
  var theme = _react.default.useContext(_terraThemeContext.default);
  var statusView = registeredStatusOverlayKeys.length > 0 && /*#__PURE__*/_react.default.createElement(_terraStatusView.default, {
    buttonAttrs: registeredButtonAttrs,
    className: cx('status-view', theme.className),
    message: registeredMessage,
    variant: registeredVariant
  });
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, customProps, {
    className: className
  }), statusView, /*#__PURE__*/_react.default.createElement("div", {
    "data-status-overlay-container-content": true,
    ref: containerRef,
    className: cx('container-content')
  }, /*#__PURE__*/_react.default.createElement(_terraScroll.default, {
    refCallback: scrollRefCallback
  }, /*#__PURE__*/_react.default.createElement(_ApplicationStatusOverlayContext.default.Provider, {
    value: contextValue
  }, children))));
};
ApplicationStatusOverlayProvider.propTypes = propTypes;
var _default = ApplicationStatusOverlayProvider;
exports.default = _default;
