// terra-content-container/lib/ContentContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraScroll = _interopRequireDefault(require("terra-scroll"));
var _ContentContainerModule = _interopRequireDefault(require("./ContentContainer.module.scss"));
var _excluded = ["header", "footer", "children", "fill", "scrollRefCallback"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var cx = _bind.default.bind(_ContentContainerModule.default);
var propTypes = {
  /**
   * The header element to be placed within the header area of the container.
   */
  header: _propTypes.default.node,
  /**
   * The footer element to be placed within the footer area of the container.
   */
  footer: _propTypes.default.node,
  /**
   * The children to be placed within the main content area of the container.
   */
  children: _propTypes.default.node,
  /**
   * Whether or not the container should expanded to fill its parent element.
   */
  fill: _propTypes.default.bool,
  /**
   * Ref callback for the scrollable area of the content container.
   */
  scrollRefCallback: _propTypes.default.func
};
var defaultProps = {
  header: undefined,
  footer: undefined,
  children: undefined,
  fill: false,
  scrollRefCallback: undefined
};
var ContentContainer = function ContentContainer(_ref) {
  var header = _ref.header,
    footer = _ref.footer,
    children = _ref.children,
    fill = _ref.fill,
    scrollRefCallback = _ref.scrollRefCallback,
    customProps = _objectWithoutProperties(_ref, _excluded);
  var contentLayoutClassNames = cx(["content-container-".concat(fill ? 'fill' : 'static'), customProps.className]);
  return /*#__PURE__*/_react.default.createElement("div", _extends({}, customProps, {
    className: contentLayoutClassNames
  }), header && /*#__PURE__*/_react.default.createElement("div", {
    className: cx('header')
  }, header), /*#__PURE__*/_react.default.createElement("div", {
    className: cx('main')
  }, /*#__PURE__*/_react.default.createElement(_terraScroll.default, {
    className: cx('normalizer'),
    refCallback: scrollRefCallback
  }, children)), footer && /*#__PURE__*/_react.default.createElement("div", {
    className: cx('footer')
  }, footer));
};
ContentContainer.propTypes = propTypes;
ContentContainer.defaultProps = defaultProps;
var _default = ContentContainer;
exports.default = _default;
