// @cerner/terra-dev-site/lib/mdx/_HTag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _MarkdownTagsModule = _interopRequireDefault(require("./MarkdownTags.module.scss"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var cx = _bind.default.bind(_MarkdownTagsModule.default);
var propTypes = {
  /**
   * The html tag for this component. Assumed to be an H tag.
   */
  Tag: _propTypes.default.string,
  /**
   * Props to apply to this component tag
   */
  props: _propTypes.default.shape({
    className: _propTypes.default.string.isRequired,
    children: _propTypes.default.node,
    id: _propTypes.default.string
  })
};

/**
 * A component to represent an H tag for MDX.
 * @param {{ Tag, props: componentProps }} props
 */
var H = function H(_ref) {
  var Tag = _ref.Tag,
    componentProps = _ref.props;
  var aRef = (0, _react.useRef)(null);
  var theme = _react.default.useContext(_terraThemeContext.default);
  // This effect is for scrolling the h-tag into view after initial load of the page.
  (0, _react.useEffect)(function () {
    if (!window.location || window.location.length < 2) {
      return;
    }
    var requestedId = window.location.hash.slice(1);
    if (componentProps.id === requestedId) {
      aRef.current.scrollIntoView();
    }
  }, [componentProps.id]);
  return /*#__PURE__*/_react.default.createElement(Tag, (0, _extends2.default)({}, componentProps, {
    className: [cx(Tag), componentProps.className].join(' ')
  }), /*#__PURE__*/_react.default.createElement("a", {
    ref: aRef,
    "aria-hidden": "true",
    href: "#".concat(componentProps.id),
    tabIndex: "-1",
    className: cx('a', 'anchor')
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: cx('icon', 'icon-link', theme.className)
  })), componentProps.children);
};
H.propTypes = propTypes;
var _default = H;
exports.default = _default;
