// eslint-plugin-formatjs/rules/enforce-description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("../util");
function checkNode(context, node) {
    var msgs = (0, util_1.extractMessages)(node, context.settings);
    var type = context.options[0];
    for (var _i = 0, msgs_1 = msgs; _i < msgs_1.length; _i++) {
        var _a = msgs_1[_i][0], description = _a.message.description, descriptionNode = _a.descriptionNode;
        if (!description) {
            if (type === 'literal' && descriptionNode) {
                context.report({
                    node: descriptionNode,
                    message: '`description` has to be a string literal (not function call or variable)',
                });
            }
            else if (!descriptionNode) {
                context.report({
                    node: node,
                    message: '`description` has to be specified in message descriptor',
                });
            }
        }
    }
}
exports.default = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Enforce description in message descriptor',
            category: 'Errors',
            recommended: false,
            url: 'https://formatjs.io/docs/tooling/linter#enforce-description',
        },
        fixable: 'code',
        schema: [
            {
                enum: ['literal', 'anything'],
            },
        ],
    },
    create: function (context) {
        var callExpressionVisitor = function (node) {
            return checkNode(context, node);
        };
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: function (node) { return checkNode(context, node); },
            CallExpression: callExpressionVisitor,
        };
    },
};
