// eslint-plugin-formatjs/rules/no-multiple-whitespaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("../util");
var MULTIPLE_SPACES = /\s{2,}/g;
function checkNode(context, node) {
    var msgs = (0, util_1.extractMessages)(node, context.settings);
    var _loop_1 = function (defaultMessage, messageNode) {
        if (!defaultMessage || !messageNode) {
            return "continue";
        }
        if (MULTIPLE_SPACES.test(defaultMessage)) {
            var reportObject = {
                node: messageNode,
                message: 'Multiple consecutive whitespaces are not allowed',
            };
            if (messageNode.type === 'Literal' && messageNode.raw) {
                reportObject.fix = function (fixer) {
                    return fixer.replaceText(messageNode, messageNode.raw.replace(MULTIPLE_SPACES, ' '));
                };
            }
            context.report(reportObject);
        }
    };
    for (var _i = 0, msgs_1 = msgs; _i < msgs_1.length; _i++) {
        var _a = msgs_1[_i][0], defaultMessage = _a.message.defaultMessage, messageNode = _a.messageNode;
        _loop_1(defaultMessage, messageNode);
    }
}
var rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow emojis in message',
            category: 'Errors',
            recommended: false,
            url: 'https://formatjs.io/docs/tooling/linter#no-emoji',
        },
        fixable: 'whitespace',
    },
    create: function (context) {
        var callExpressionVisitor = function (node) {
            return checkNode(context, node);
        };
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: function (node) { return checkNode(context, node); },
            CallExpression: callExpressionVisitor,
        };
    },
};
exports.default = rule;
