// eslint-plugin-formatjs/rules/no-multiple-plurals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_1 = require("../util");
var icu_messageformat_parser_1 = require("@formatjs/icu-messageformat-parser");
var MultiplePlurals = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(MultiplePlurals, _super);
    function MultiplePlurals() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.message = 'Cannot specify more than 1 plural rules';
        return _this;
    }
    return MultiplePlurals;
}(Error));
function verifyAst(ast, pluralCount) {
    if (pluralCount === void 0) { pluralCount = { count: 0 }; }
    for (var _i = 0, ast_1 = ast; _i < ast_1.length; _i++) {
        var el = ast_1[_i];
        if ((0, icu_messageformat_parser_1.isPluralElement)(el)) {
            pluralCount.count++;
            if (pluralCount.count > 1) {
                throw new MultiplePlurals();
            }
            var options = el.options;
            for (var _a = 0, _b = Object.keys(options); _a < _b.length; _a++) {
                var selector = _b[_a];
                verifyAst(options[selector].value, pluralCount);
            }
        }
    }
}
function checkNode(context, node) {
    var msgs = (0, util_1.extractMessages)(node, context.settings);
    for (var _i = 0, msgs_1 = msgs; _i < msgs_1.length; _i++) {
        var _a = msgs_1[_i][0], defaultMessage = _a.message.defaultMessage, messageNode = _a.messageNode;
        if (!defaultMessage || !messageNode) {
            continue;
        }
        try {
            verifyAst((0, icu_messageformat_parser_1.parse)(defaultMessage, {
                ignoreTag: context.settings.ignoreTag,
            }));
        }
        catch (e) {
            context.report({
                node: messageNode,
                message: e instanceof Error ? e.message : String(e),
            });
        }
    }
}
var rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow multiple plural rules in the same message',
            category: 'Errors',
            recommended: false,
            url: 'https://formatjs.io/docs/tooling/linter#no-multiple-plurals',
        },
        fixable: 'code',
    },
    create: function (context) {
        var callExpressionVisitor = function (node) {
            return checkNode(context, node);
        };
        if (context.parserServices.defineTemplateBodyVisitor) {
            return context.parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: function (node) { return checkNode(context, node); },
            CallExpression: callExpressionVisitor,
        };
    },
};
exports.default = rule;
