// terra-application/lib/terra-dev-site/test/application-status-overlay/StatusOverlayAllProps.test.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _ApplicationStatusOverlay = _interopRequireDefault(require("../../../../lib/application-status-overlay/ApplicationStatusOverlay"));
var _ApplicationStatusOverlayProvider = _interopRequireDefault(require("../../../../lib/application-status-overlay/ApplicationStatusOverlayProvider"));
var ApplicationStatusOverlayAllProps = function ApplicationStatusOverlayAllProps() {
  var StatusViewButtons = [{
    text: 'Action 1',
    key: 1
  }, {
    text: 'Action 2',
    key: 2
  }];
  return (
    /*#__PURE__*/
    // eslint-disable-next-line react/forbid-dom-props
    _react.default.createElement("div", {
      style: {
        height: '500px'
      }
    }, /*#__PURE__*/_react.default.createElement(_ApplicationStatusOverlayProvider.default, null, /*#__PURE__*/_react.default.createElement(_ApplicationStatusOverlay.default, {
      buttonAttrs: StatusViewButtons,
      message: "Status View with all props",
      variant: "no-data"
    })))
  );
};
var _default = ApplicationStatusOverlayAllProps;
exports.default = _default;
