// terra-application/lib/workspace/overlays/WorkspaceContentActivityOverlay.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _reactIntl = require("react-intl");
var _IconSpinner = _interopRequireDefault(require("terra-icon/lib/icon/IconSpinner"));
var _theme = require("../../theme");
var _WorkspaceContentActivityOverlayModule = _interopRequireDefault(require("./WorkspaceContentActivityOverlay.module.scss"));
var cx = _bind.default.bind(_WorkspaceContentActivityOverlayModule.default);
var propTypes = {
  /**
   * String determining which activity variant to render.
   */
  variant: _propTypes.default.oneOf(['default', 'loading'])
};
var defaultProps = {
  variant: 'default'
};
var WorkspaceContentActivityOverlay = function WorkspaceContentActivityOverlay(_ref) {
  var variant = _ref.variant;
  var theme = _react.default.useContext(_theme.ThemeContext);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('activity-overlay', theme.className),
    role: "status",
    "data-testid": "workspace-content-activity-overlay"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cx('content')
  }, /*#__PURE__*/_react.default.createElement(_IconSpinner.default, {
    className: cx('icon'),
    isSpin: true,
    height: "36",
    width: "36"
  }), variant === 'loading' ? /*#__PURE__*/_react.default.createElement(_reactIntl.FormattedMessage, {
    id: "terraApplication.workspace.activityOverlay.loading"
  }, function (loadingMessage) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: cx('message')
    }, loadingMessage);
  }) : undefined));
};
WorkspaceContentActivityOverlay.propTypes = propTypes;
WorkspaceContentActivityOverlay.defaultProps = defaultProps;
var _default = WorkspaceContentActivityOverlay;
exports.default = _default;
