// @cerner/terra-dev-site/lib/site/_DevSiteApplicationBase.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactRouterDom = require("react-router-dom");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _terraApplication = _interopRequireDefault(require("terra-application"));
var _AppSettingsContext = _interopRequireDefault(require("./_AppSettingsContext"));
var propTypes = {
  /**
   * The remainder of the application to render.
   */
  children: _propTypes.default.node,
  /**
   * An override component for a custom application base. Must render children and adhere to the terra application base api.
   */
  ApplicationBase: _propTypes.default.func
};
var DevSiteApplicationContainer = function DevSiteApplicationContainer(_ref) {
  var children = _ref.children,
    ApplicationBase = _ref.ApplicationBase;
  var isRaw = (0, _reactRouterDom.useRouteMatch)('/raw');
  var _React$useContext = _react.default.useContext(_AppSettingsContext.default),
    currentLocale = _React$useContext.currentLocale,
    currentThemeClassName = _React$useContext.currentThemeClassName;
  var AppBase = ApplicationBase || _terraApplication.default;
  return /*#__PURE__*/_react.default.createElement(AppBase, {
    locale: currentLocale,
    themeName: currentThemeClassName,
    themeIsGlobal: true,
    unloadPromptIsDisabled: !!isRaw
  }, children);
};
DevSiteApplicationContainer.propTypes = propTypes;
var _default = DevSiteApplicationContainer;
exports.default = _default;
