// @cerner/terra-dev-site/lib/site/Site.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactRouterDom = require("react-router-dom");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _modalManager = _interopRequireDefault(require("terra-application/lib/modal-manager"));
var _DevSiteLayout = _interopRequireDefault(require("../layouts/_DevSiteLayout"));
var _AppSettingsProvider = _interopRequireDefault(require("./_AppSettingsProvider"));
var _siteConfigShapes = _interopRequireDefault(require("./siteConfigShapes"));
var _TerraMdxProvider = _interopRequireDefault(require("../mdx/_TerraMdxProvider"));
var _DevSiteRouter = _interopRequireDefault(require("./_DevSiteRouter"));
var _DevSiteApplicationBase = _interopRequireDefault(require("./_DevSiteApplicationBase"));
var propTypes = {
  /**
   * The site config for the application.
   */
  siteConfig: _siteConfigShapes.default.isRequired,
  /**
   * The component representing the providers layer of terra-dev-site. Must render children.
   */
  Providers: _propTypes.default.func,
  /**
   * An override component for a custom application base. Must render children and adhere to the terra application base api.
   */
  ApplicationBase: _propTypes.default.func,
  /**
   * An override component for a custom application base. Must render children and adhere to the terra application base api.
   */
  ApplicationNavigation: _propTypes.default.func
};
var Site = function Site(_ref) {
  var siteConfig = _ref.siteConfig,
    Providers = _ref.Providers,
    ApplicationBase = _ref.ApplicationBase,
    ApplicationNavigation = _ref.ApplicationNavigation;
  return /*#__PURE__*/_react.default.createElement(_AppSettingsProvider.default, {
    settingsConfig: siteConfig.settingsConfig
  }, /*#__PURE__*/_react.default.createElement(_reactRouterDom.BrowserRouter, {
    basename: siteConfig.basename
  }, /*#__PURE__*/_react.default.createElement(_DevSiteRouter.default, {
    sites: siteConfig.sites,
    routesMap: siteConfig.routesMap
  }, /*#__PURE__*/_react.default.createElement(_DevSiteApplicationBase.default, {
    ApplicationBase: ApplicationBase
  }, /*#__PURE__*/_react.default.createElement(_TerraMdxProvider.default, null, /*#__PURE__*/_react.default.createElement(_modalManager.default, null, Providers ? /*#__PURE__*/_react.default.createElement(Providers, null, /*#__PURE__*/_react.default.createElement(_DevSiteLayout.default, {
    siteConfig: siteConfig,
    ApplicationNavigation: ApplicationNavigation
  })) : /*#__PURE__*/_react.default.createElement(_DevSiteLayout.default, {
    siteConfig: siteConfig,
    ApplicationNavigation: ApplicationNavigation
  })))))));
};
Site.propTypes = propTypes;
var _default = Site;
exports.default = _default;
