// @cerner/terra-dev-site/lib/layouts/_ContentLayout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _reactRouterDom = require("react-router-dom");
var _terraImage = _interopRequireDefault(require("terra-image"));
var _Suspense = _interopRequireDefault(require("../shared/Suspense"));
var _ContentLoaded = _interopRequireDefault(require("../content/_ContentLoaded"));
var _siteConfigShapes = require("../site/siteConfigShapes");
var _kaiju = _interopRequireDefault(require("../pages/kaiju-404.gif"));
var propTypes = {
  /**
   * Config describing the page
   */
  pageContentConfig: _siteConfigShapes.pageContentConfigShape.isRequired,
  /**
   * Function called to request closing the modal
   */
  contentImports: _siteConfigShapes.contentImportsShape.isRequired
};
var ContentLayout = function ContentLayout(_ref) {
  var pageContentConfig = _ref.pageContentConfig,
    contentImports = _ref.contentImports;
  var location = (0, _reactRouterDom.useLocation)();
  var pathname = location.pathname;
  var ContentComponent = contentImports[pathname];
  var _React$useState = _react.default.useState(),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    loadingFailed = _React$useState2[0],
    setLoadingFailed = _React$useState2[1];
  if (!pageContentConfig || !ContentComponent) {
    return /*#__PURE__*/_react.default.createElement(_terraImage.default, {
      src: _kaiju.default,
      width: "100%",
      alt: "404"
    });
  }
  if (loadingFailed) {
    return /*#__PURE__*/_react.default.createElement(_ContentLoaded.default, {
      type: pageContentConfig.type,
      isScrollContainer: true
    }, /*#__PURE__*/_react.default.createElement("div", null, "Error: chunk failed to load."));
  }
  return /*#__PURE__*/_react.default.createElement(_Suspense.default, {
    onError: function onError() {
      return setLoadingFailed(true);
    }
  }, /*#__PURE__*/_react.default.createElement(_ContentLoaded.default, {
    type: pageContentConfig.type,
    isScrollContainer: true
  }, /*#__PURE__*/_react.default.createElement(ContentComponent, null)));
};
ContentLayout.propTypes = propTypes;
var _default = ContentLayout;
exports.default = _default;
