// @cerner/terra-dev-site/lib/layouts/_RawLayout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _reactRouterDom = require("react-router-dom");
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraImage = _interopRequireDefault(require("terra-image"));
var _ContentLoaded = _interopRequireDefault(require("../content/_ContentLoaded"));
var _Suspense = _interopRequireDefault(require("../shared/Suspense"));
var _kaiju = _interopRequireDefault(require("../pages/kaiju-404.gif"));
var _siteConfigShapes = _interopRequireDefault(require("../site/siteConfigShapes"));
var _ContentLayoutModule = _interopRequireDefault(require("./ContentLayout.module.scss"));
var cx = _bind.default.bind(_ContentLayoutModule.default);
var propTypes = {
  /**
   * The site config for the application
   */
  siteConfig: _siteConfigShapes.default.isRequired
};
var Raw = function Raw(_ref) {
  var siteConfig = _ref.siteConfig;
  var location = (0, _reactRouterDom.useLocation)();
  var pathname = location.pathname.substring(4).toLocaleLowerCase();
  var pageContentConfig = siteConfig.pageConfig[pathname];
  var ContentComponent = siteConfig.contentImports[pathname];
  var _React$useState = _react.default.useState(),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    loadingFailed = _React$useState2[0],
    setLoadingFailed = _React$useState2[1];
  if (!pageContentConfig || !ContentComponent) {
    return /*#__PURE__*/_react.default.createElement(_terraImage.default, {
      src: _kaiju.default,
      width: "100%",
      alt: "404"
    });
  }
  if (loadingFailed) {
    return /*#__PURE__*/_react.default.createElement(_ContentLoaded.default, {
      type: pageContentConfig.type,
      isScrollContainer: true
    }, /*#__PURE__*/_react.default.createElement("div", null, "Error: chunk failed to load."));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    role: "main",
    className: cx('main')
  }, /*#__PURE__*/_react.default.createElement(_Suspense.default, {
    fallback: /*#__PURE__*/_react.default.createElement("div", null, "loading"),
    onError: function onError() {
      return setLoadingFailed(true);
    }
  }, /*#__PURE__*/_react.default.createElement(_ContentLoaded.default, {
    type: pageContentConfig.type,
    isScrollContainer: true
  }, /*#__PURE__*/_react.default.createElement(ContentComponent, null))));
};
Raw.propTypes = propTypes;
var _default = Raw;
exports.default = _default;
