// terra-list/lib/ListSection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _ListSectionHeader = _interopRequireDefault(require("./ListSectionHeader"));
var _ListModule = _interopRequireDefault(require("./List.module.scss"));
var _excluded = ["children", "isCollapsed", "isCollapsible"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var cx = _bind.default.bind(_ListModule.default);
var propTypes = {
  /**
   * The children list items passed to the component.
   */
  children: _propTypes.default.node,
  /**
   * Whether or not the section is collapsed.
   */
  isCollapsed: _propTypes.default.bool,
  /**
   * Whether or not the section can be collapsed.
   */
  isCollapsible: _propTypes.default.bool,
  /**
   * Optionally sets the heading level. One of `1`, `2`, `3`, `4`, `5`, `6`.
   */
  level: _propTypes.default.oneOf([1, 2, 3, 4, 5, 6]),
  /**
   * The associated metaData to be provided in the onSelect callback.
   */
  // eslint-disable-next-line react/forbid-prop-types
  metaData: _propTypes.default.object,
  /**
   * Function callback for when the appropriate click or key action is performed.
   * Callback contains the javascript event and prop metadata, e.g. onSelect(event, metaData)
   */
  onSelect: _propTypes.default.func,
  /**
   * Function callback passthrough for the ref of the section li.
   */
  refCallback: _propTypes.default.func,
  /**
   * Title text to be placed within the section header.
   */
  title: _propTypes.default.string.isRequired
};
var defaultProps = {
  children: [],
  isCollapsed: false,
  isCollapsible: false,
  level: 1
};
var ListSection = function ListSection(_ref) {
  var children = _ref.children,
    isCollapsed = _ref.isCollapsed,
    isCollapsible = _ref.isCollapsible,
    customProps = _objectWithoutProperties(_ref, _excluded);
  var sectionItems;
  if (!isCollapsible || !isCollapsed) {
    sectionItems = children;
  }
  var theme = _react.default.useContext(_terraThemeContext.default);
  var listClassNames = (0, _classnames.default)(cx('list', theme.className), customProps.className);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ListSectionHeader.default, _extends({}, customProps, {
    isCollapsible: isCollapsible,
    isCollapsed: isCollapsed
  })), /*#__PURE__*/_react.default.createElement("li", null, /*#__PURE__*/_react.default.createElement("ul", _extends({}, customProps, {
    className: listClassNames
  }), sectionItems)));
};
ListSection.propTypes = propTypes;
ListSection.defaultProps = defaultProps;
var _default = ListSection;
exports.default = _default;
