// @cerner/terra-dev-site/lib/layouts/page-container/PageContainerContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.contextShape = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
/**
 * The PageContainerContext is used by PageContainers to communicate state
 * and configuration to Page components rendered within them.
 */
var PageContainerContext = /*#__PURE__*/_react.default.createContext();
var actionShape = _propTypes.default.shape({
  /**
   * An identifier for the action. This must be unique within the provided
   * actions.
   */
  key: _propTypes.default.string.isRequired,
  /**
   * A string description for the action. This will accessible to users and
   * should be translated sif necessary.
   */
  label: _propTypes.default.string.isRequired,
  /**
   * A graphic representing the action. This may be used by the context consumer
   * to render controls for the action.
   */
  icon: _propTypes.default.elementType,
  /**
   * A function executed upon selection of the action.
   */
  onSelect: _propTypes.default.func
});
var contextShape = {
  /**
   * Action definitions to be rendered by the Page within the Page's leading
   * action region.
   */
  containerStartActions: _propTypes.default.arrayOf(actionShape),
  /**
   * Action definitions to be rendered by the Page within the Page's trailing
   * action region.
   */
  containerEndActions: _propTypes.default.arrayOf(actionShape)
};
exports.contextShape = contextShape;
var _default = PageContainerContext;
exports.default = _default;
