// @cerner/terra-dev-site/lib/pages/page/overlays/PageActivityOverlay.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _reactIntl = require("react-intl");
var _IconSpinner = _interopRequireDefault(require("terra-icon/lib/icon/IconSpinner"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _PageActivityOverlayModule = _interopRequireDefault(require("./PageActivityOverlay.module.scss"));
var cx = _bind.default.bind(_PageActivityOverlayModule.default);
var propTypes = {
  variant: _propTypes.default.oneOf(['default', 'loading'])
};
var defaultProps = {
  variant: 'default'
};
var PageActivityOverlay = function PageActivityOverlay(_ref) {
  var variant = _ref.variant;
  var theme = _react.default.useContext(_terraThemeContext.default);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('activity-overlay', theme.className),
    "aria-live": "polite"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cx('content')
  }, /*#__PURE__*/_react.default.createElement(_IconSpinner.default, {
    className: cx('icon'),
    isSpin: true,
    height: "36",
    width: "36"
  }), variant === 'loading' ? /*#__PURE__*/_react.default.createElement(_reactIntl.FormattedMessage, {
    id: "terraDevSite.pageActivityOverlayContainer.loading"
  }, function (loadingMessage) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: cx('message')
    }, loadingMessage);
  }) : undefined));
};
PageActivityOverlay.propTypes = propTypes;
PageActivityOverlay.defaultProps = defaultProps;
var _default = PageActivityOverlay;
exports.default = _default;
