// terra-form-select/lib/shared/_NoResults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _reactIntl = require("react-intl");
var _NoResultsModule = _interopRequireDefault(require("./_NoResults.module.scss"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var cx = _bind.default.bind(_NoResultsModule.default);
var propTypes = {
  /**
   * Content to display when no results are found.
   */
  noResultContent: _propTypes.default.node,
  /**
   * The filter value.
   */
  value: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
};
var NoResults = function NoResults(_ref) {
  var noResultContent = _ref.noResultContent,
    value = _ref.value;
  var theme = _react.default.useContext(_terraThemeContext.default);
  return /*#__PURE__*/_react.default.createElement("div", {
    role: "option",
    className: cx('no-results', theme.className),
    "aria-selected": "false"
  }, noResultContent || /*#__PURE__*/_react.default.createElement(_reactIntl.FormattedMessage, {
    id: "Terra.form.select.noResults",
    values: {
      text: value
    }
  }));
};
NoResults.propTypes = propTypes;
var _default = NoResults;
exports.default = _default;
