// terra-form-select/lib/shared/_OptGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _reactIntl = require("react-intl");
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _OptGroupModule = _interopRequireDefault(require("./_OptGroup.module.scss"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var cx = _bind.default.bind(_OptGroupModule.default);
var propTypes = {
  /**
   * The group options.
   */
  children: _propTypes.default.node,
  /**
   * Whether the option group is disabled.
   */
  disabled: _propTypes.default.bool,
  /**
   * @private
   * The intl object containing translations. This is retrieved from the context automatically by injectIntl.
   */
  intl: _propTypes.default.shape({
    formatMessage: _propTypes.default.func
  }).isRequired,
  /**
   * The group label.
   */
  label: _propTypes.default.node.isRequired
};
var defaultProps = {
  disabled: false
};
var OptGroup = function OptGroup(_ref) {
  var children = _ref.children,
    disabled = _ref.disabled,
    intl = _ref.intl,
    label = _ref.label;
  var theme = _react.default.useContext(_terraThemeContext.default);
  return /*#__PURE__*/_react.default.createElement("li", {
    className: cx('opt-group', {
      'is-disabled': disabled
    }, theme.className),
    role: "option",
    "aria-selected": "false"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cx('label')
  }, label), /*#__PURE__*/_react.default.createElement("ul", {
    className: cx('options'),
    role: "listbox",
    "aria-label": intl.formatMessage({
      id: 'Terra.form.select.option'
    })
  }, _react.default.Children.map(children, function (child) {
    return /*#__PURE__*/_react.default.cloneElement(child, {
      disabled: disabled || !!child.props.disabled
    });
  })));
};
OptGroup.propTypes = propTypes;
OptGroup.defaultProps = defaultProps;
OptGroup.isOptGroup = true;
var _default = (0, _reactIntl.injectIntl)(OptGroup);
exports.default = _default;
