// terra-form-select/lib/shared/_Tag.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _TagModule = _interopRequireDefault(require("./_Tag.module.scss"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var cx = _bind.default.bind(_TagModule.default);
var propTypes = {
  /**
   * The content of the tag.
   */
  children: _propTypes.default.node.isRequired,
  /**
   * Callback function triggered when the tag is deselected.
   */
  onDeselect: _propTypes.default.func.isRequired,
  /**
   * The value of the tag.
   */
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]).isRequired
};

/* eslint-disable jsx-a11y/no-static-element-interactions */
var Tag = function Tag(_ref) {
  var children = _ref.children,
    onDeselect = _ref.onDeselect,
    value = _ref.value;
  var theme = _react.default.useContext(_terraThemeContext.default);
  return /*#__PURE__*/_react.default.createElement("li", {
    className: cx('tag', theme.className)
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: cx('display')
  }, children), /*#__PURE__*/_react.default.createElement("span", {
    className: cx('deselect'),
    onClick: function onClick() {
      onDeselect(value);
    },
    role: "presentation"
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: cx('icon')
  })));
};
Tag.propTypes = propTypes;
var _default = Tag;
exports.default = _default;
