// terra-application/lib/terra-dev-site/test/workspace/MockContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _applicationContainer = require("../../../application-container");
var _NotificationBanner = _interopRequireDefault(require("../../../notification-banner/NotificationBanner"));
var _TestStylesModule = _interopRequireDefault(require("./TestStyles.module.scss"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var cx = _bind.default.bind(_TestStylesModule.default);
var propTypes = {
  initialCount: _propTypes.default.number,
  title: _propTypes.default.string,
  onShowActivityOverlay: _propTypes.default.func,
  onShowStatusOverlay: _propTypes.default.func
};
var MockContent = function MockContent(_ref) {
  var _ref$initialCount = _ref.initialCount,
    initialCount = _ref$initialCount === void 0 ? 0 : _ref$initialCount,
    _ref$title = _ref.title,
    title = _ref$title === void 0 ? '' : _ref$title,
    onShowActivityOverlay = _ref.onShowActivityOverlay,
    onShowStatusOverlay = _ref.onShowStatusOverlay;
  var activeMainPage = (0, _applicationContainer.useActiveMainPage)();
  var _useState = (0, _react.useState)(initialCount),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    clickCount = _useState2[0],
    setClickCount = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
    showAlertBanner = _useState4[0],
    setShowAlertBanner = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
    showWarningBanner = _useState6[0],
    setShowWarningBanner = _useState6[1];
  var _useState7 = (0, _react.useState)(false),
    _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
    showErrorBanner = _useState8[0],
    setShowErrorBanner = _useState8[1];
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('content-container')
  }, /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      return setClickCount(clickCount + 1);
    }
  }, "Increment Counter"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    "data-testid": "test-workspace-banner-button",
    onClick: function onClick() {
      return setShowAlertBanner(true);
    }
  }, "Show Alert Banner"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    "data-testid": "test-workspace-banner-warning-button",
    onClick: function onClick() {
      return setShowWarningBanner(true);
    }
  }, "Show Warning Banner"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    "data-testid": "test-workspace-banner-error-button",
    onClick: function onClick() {
      return setShowErrorBanner(true);
    }
  }, "Show Error Banner"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      onShowActivityOverlay(true);
      setTimeout(function () {
        onShowActivityOverlay(false);
      }, 5000);
    }
  }, "Show Activity Overlay"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: function onClick() {
      onShowStatusOverlay(true);
      setTimeout(function () {
        onShowStatusOverlay(false);
      }, 5000);
    }
  }, "Show Status Overlay"), showAlertBanner && /*#__PURE__*/_react.default.createElement(_NotificationBanner.default, {
    variant: "hazard-high",
    id: "workspace-test-alert-banner",
    description: "There is something very wrong",
    onRequestClose: function onRequestClose() {
      return setShowAlertBanner(false);
    }
  }), showWarningBanner && /*#__PURE__*/_react.default.createElement(_NotificationBanner.default, {
    variant: "hazard-medium",
    id: "workspace-test-warning-banner",
    description: "There is something kind of wrong",
    onRequestClose: function onRequestClose() {
      return setShowWarningBanner(false);
    }
  }), showErrorBanner && /*#__PURE__*/_react.default.createElement(_NotificationBanner.default, {
    variant: "error",
    id: "workspace-test-error-banner",
    description: "Something broke",
    onRequestClose: function onRequestClose() {
      return setShowErrorBanner(false);
    },
    bannerAction: {
      text: 'Help',
      onClick: function onClick() {}
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    "aria-live": "polite"
  }, /*#__PURE__*/_react.default.createElement("p", null, "".concat(title, "'s Click Counter: ").concat(clickCount))), /*#__PURE__*/_react.default.createElement("p", null, "Active Main Page Key:", ' ', activeMainPage === null || activeMainPage === void 0 ? void 0 : activeMainPage.pageKey), /*#__PURE__*/_react.default.createElement("p", null, "Active Main Page Label:", ' ', activeMainPage === null || activeMainPage === void 0 ? void 0 : activeMainPage.pageLabel), /*#__PURE__*/_react.default.createElement("p", null, "Active Main Page MetaData:", ' ', "".concat(JSON.stringify(activeMainPage === null || activeMainPage === void 0 ? void 0 : activeMainPage.pageMetaData))), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam sit amet lacus cursus massa ullamcorper scelerisque. Aenean vitae posuere neque, consequat dapibus diam. Proin convallis venenatis magna, sit amet volutpat erat. Nulla sodales eu est sit amet sagittis. Suspendisse lacinia diam ut justo venenatis, a sollicitudin ante venenatis. Vivamus a leo ullamcorper, tristique diam id, vestibulum felis. Nullam mattis eget eros vestibulum porttitor. Duis eget massa nec urna ultrices laoreet. Aenean rhoncus mauris in luctus blandit. Morbi tempor enim a libero placerat, at bibendum elit luctus. In in tempor neque, laoreet facilisis quam. Fusce faucibus dui eget erat gravida egestas. Nullam laoreet purus eget urna placerat, sit amet ultrices mi tristique. Sed vulputate gravida risus, vehicula rhoncus ipsum tempus id. Phasellus aliquet nec mi non pretium. Maecenas turpis nulla, mollis et rhoncus vel, porttitor id nisl."), /*#__PURE__*/_react.default.createElement("p", null, "In vitae tortor elementum, venenatis sem id, condimentum mauris. Mauris eget eros auctor, ultrices lacus imperdiet, aliquet lorem. Nulla facilisi. Sed libero odio, tincidunt eget congue sed, ornare at nulla. In feugiat, felis a mattis viverra, quam massa rhoncus tortor, sit amet luctus augue dolor bibendum metus. Aliquam eget malesuada est. Aenean at mi ut augue placerat vulputate. Quisque vulputate iaculis semper. Ut vestibulum interdum quam, a facilisis purus dapibus sit amet. Curabitur tempus neque at tortor ultricies mollis. Aliquam commodo pulvinar lectus, sollicitudin tincidunt nulla finibus non."), /*#__PURE__*/_react.default.createElement("p", null, "Donec ornare nisi ut lectus luctus, vitae venenatis tellus molestie. Quisque augue ante, pulvinar eget ullamcorper eget, pharetra in est. Aenean urna turpis, lobortis in egestas eu, blandit eget sapien. Suspendisse vitae iaculis justo. Sed ex turpis, ultrices ut lobortis a, vestibulum a tortor. Nullam ut risus ex. Duis condimentum dui sem, interdum porttitor lectus scelerisque nec. Curabitur suscipit felis in porta molestie."), /*#__PURE__*/_react.default.createElement("p", null, "Etiam non malesuada urna. Vivamus posuere orci augue, nec egestas mauris cursus blandit. Maecenas eu nisl sed nulla venenatis tincidunt. Donec at cursus sapien, nec scelerisque nibh. Vivamus suscipit consequat ex, id feugiat nisl iaculis id. Pellentesque tristique sodales turpis. Cras rutrum nibh laoreet libero dictum, quis condimentum arcu porttitor. Donec ac aliquam orci. Vestibulum luctus sed mi tincidunt volutpat. Mauris ut neque pellentesque, volutpat ex vel, efficitur massa. Phasellus hendrerit mattis semper. Morbi mi nulla, tincidunt at auctor euismod, viverra at libero."), /*#__PURE__*/_react.default.createElement("p", null, "Maecenas aliquam congue augue sed tristique. Vivamus eu consectetur sapien, in eleifend magna. Etiam laoreet ligula eu fermentum tincidunt. Interdum et malesuada fames ac ante ipsum primis in faucibus. Sed magna nunc, tempor vitae sagittis eu, vulputate ac lorem. Etiam massa ante, aliquam ac consequat et, efficitur dictum turpis. Integer malesuada purus in lobortis ornare. In hac habitasse platea dictumst. Aliquam finibus ipsum quis nibh volutpat, a posuere mi vestibulum. Cras tincidunt odio et aliquet mollis. Ut ac felis neque. Integer nulla elit, mattis eleifend dui pretium, euismod tristique nulla. Curabitur vulputate viverra odio, vitae blandit est dapibus eu."));
};
MockContent.propTypes = propTypes;
var _default = MockContent;
exports.default = _default;
