// terra-application/lib/terra-dev-site/test/workspace/OverlayWorkspace.test.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _workspace = _interopRequireWildcard(require("../../../workspace"));
var _ActiveMainPageContext = _interopRequireDefault(require("../../../application-container/private/active-main-page/ActiveMainPageContext"));
var _Tab = _interopRequireDefault(require("./Tab1"));
var _Tab2 = _interopRequireDefault(require("./Tab2"));
var _Tab3 = _interopRequireDefault(require("./Tab3"));
var _Tab4 = _interopRequireDefault(require("./Tab4"));
var _Tab5 = _interopRequireDefault(require("./Tab5"));
var _Tab6 = _interopRequireDefault(require("./Tab6"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var sizeMap = {
  small: '320px',
  medium: '500px',
  large: '1000px'
};
var WorkspaceTest = function WorkspaceTest() {
  var _React$useState = _react.default.useState('tab-1'),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    activeItemKey = _React$useState2[0],
    setActiveItemKey = _React$useState2[1];
  var _React$useState3 = _react.default.useState('large'),
    _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
    workspaceSize = _React$useState4[0],
    setWorkspaceSize = _React$useState4[1];
  var activeMainPageRef = _react.default.useRef({
    pageKey: 'page-1',
    pageLabel: 'Test Page',
    pageMetaData: {
      data: 'data here'
    }
  });
  var onRequestSizeChange = function onRequestSizeChange(size) {
    setWorkspaceSize(size);
  };
  var onRequestClose = function onRequestClose() {
    console.log('onRequestClose'); // eslint-disable-line no-console
  };

  return /*#__PURE__*/_react.default.createElement(_ActiveMainPageContext.default.Provider, {
    value: activeMainPageRef.current
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      height: '500px',
      width: sizeMap[workspaceSize]
    } // eslint-disable-line react/forbid-dom-props
  }, /*#__PURE__*/_react.default.createElement(_workspace.default, {
    id: "overlay-test-id",
    dismissButtonIsVisible: true,
    isPresentedAsOverlay: true,
    activeItemKey: activeItemKey,
    onRequestActivate: function onRequestActivate(key) {
      return setActiveItemKey(key);
    },
    activeSize: workspaceSize,
    sizeOptions: [{
      key: 'small',
      text: 'Small'
    }, {
      key: 'medium',
      text: 'Medium'
    }, {
      key: 'large',
      text: 'Large'
    }],
    onRequestSizeChange: onRequestSizeChange,
    onRequestDismiss: onRequestClose
  }, /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceItem, {
    itemKey: "tab-1",
    label: "Tab 1",
    render: function render() {
      return /*#__PURE__*/_react.default.createElement(_Tab.default, null);
    }
  }), /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceItem, {
    itemKey: "tab-2",
    label: "Tab 2",
    render: function render() {
      return /*#__PURE__*/_react.default.createElement(_Tab2.default, null);
    }
  }), /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceItem, {
    itemKey: "tab-3",
    label: "Tab 3",
    render: function render() {
      return /*#__PURE__*/_react.default.createElement(_Tab3.default, null);
    }
  }), /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceItem, {
    itemKey: "tab-4",
    label: "Tab 4",
    render: function render() {
      return /*#__PURE__*/_react.default.createElement(_Tab4.default, null);
    }
  }), /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceItem, {
    itemKey: "tab-5",
    label: "Tab 5",
    render: function render() {
      return /*#__PURE__*/_react.default.createElement(_Tab5.default, null);
    }
  }), /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceItem, {
    itemKey: "tab-6",
    label: "Tab 6",
    render: function render() {
      return /*#__PURE__*/_react.default.createElement(_Tab6.default, null);
    }
  }))));
};
var _default = WorkspaceTest;
exports.default = _default;
