// terra-application/lib/terra-dev-site/test/workspace/Tab2.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _workspace = require("../../../workspace");
var _MockContent = _interopRequireDefault(require("./MockContent"));
var Tab2 = function Tab2() {
  var _React$useState = _react.default.useState(),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    isLoading = _React$useState2[0],
    setIsLoading = _React$useState2[1];
  var _React$useState3 = _react.default.useState(),
    _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
    showStatusOverlay = _React$useState4[0],
    setShowStatusOverlay = _React$useState4[1];
  return /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent, {
    activityOverlay: isLoading ? /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.ActivityOverlay, {
      variant: "loading"
    }) : undefined,
    statusOverlay: showStatusOverlay ? /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.StatusOverlay, {
      variant: "not-authorized"
    }) : undefined
  }, /*#__PURE__*/_react.default.createElement(_MockContent.default, {
    title: "Tab 2",
    onShowActivityOverlay: setIsLoading,
    onShowStatusOverlay: setShowStatusOverlay
  }));
};
var _default = Tab2;
exports.default = _default;
