// terra-application/lib/terra-dev-site/test/workspace/Tab3.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _terraToolbar = _interopRequireDefault(require("terra-toolbar"));
var _terraButton = _interopRequireDefault(require("terra-button"));
var _IconEdit = _interopRequireDefault(require("terra-icon/lib/icon/IconEdit"));
var _IconAdd = _interopRequireDefault(require("terra-icon/lib/icon/IconAdd"));
var _workspace = require("../../../workspace");
var _MockContent = _interopRequireDefault(require("./MockContent"));
var Tab3 = function Tab3() {
  var _React$useState = _react.default.useState(),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    isLoading = _React$useState2[0],
    setIsLoading = _React$useState2[1];
  var _React$useState3 = _react.default.useState(),
    _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
    showStatusOverlay = _React$useState4[0],
    setShowStatusOverlay = _React$useState4[1];
  return /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent, {
    toolbar: /*#__PURE__*/_react.default.createElement(_terraToolbar.default, null, /*#__PURE__*/_react.default.createElement(_terraButton.default, {
      text: "Edit",
      variant: "utility",
      icon: /*#__PURE__*/_react.default.createElement(_IconEdit.default, null)
    }), /*#__PURE__*/_react.default.createElement(_terraButton.default, {
      text: "Add",
      variant: "utility",
      icon: /*#__PURE__*/_react.default.createElement(_IconAdd.default, null)
    })),
    activityOverlay: isLoading ? /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.ActivityOverlay, {
      variant: "loading"
    }) : undefined,
    statusOverlay: showStatusOverlay ? /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.StatusOverlay, {
      variant: "not-authorized"
    }) : undefined
  }, /*#__PURE__*/_react.default.createElement(_MockContent.default, {
    title: "Tab 3",
    onShowActivityOverlay: setIsLoading,
    onShowStatusOverlay: setShowStatusOverlay
  }));
};
var _default = Tab3;
exports.default = _default;
