// terra-application/lib/terra-dev-site/test/workspace/Tab5.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireDefault(require("react"));
var _workspace = require("../../../workspace");
var Tab5 = function Tab5() {
  var _React$useState = _react.default.useState(),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    isLoading = _React$useState2[0],
    setIsLoading = _React$useState2[1];
  return /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent, {
    activityOverlay: isLoading ? /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.ActivityOverlay, {
      variant: "loading"
    }) : undefined,
    statusOverlay: /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.StatusOverlay, {
      variant: "error",
      message: "This is an error message."
    }, /*#__PURE__*/_react.default.createElement(_workspace.WorkspaceContent.StatusOverlay.Button, {
      text: "Retry",
      onClick: function onClick() {
        setIsLoading(true);
        setTimeout(function () {
          setIsLoading(false);
        }, 5000);
      }
    }))
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, "Tab 5 Content")));
};
var _default = Tab5;
exports.default = _default;
