// override-require/dist/overrideRequire.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _module = require('module');

var _module2 = _interopRequireDefault(_module);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*:: type ParentType = {
  children: Array<any>,
  exports: any,
  filename: string,
  id: string,
  loaded: boolean,
  parent: any,
  paths: Array<string>
};*/ // @flow

/*:: type IsOverrideType = (request: string, parent: ParentType) => boolean;*/
/*:: type ResolverOverrideType = (request: string, parent: ParentType) => any;*/


/**
 * @param isOverride A condition used to check whether to override Module._load.
 * @param resolveOverride A function used to override Module._load result.
 */
/*:: type OverrideRequireType = (...rest: Array<void>) => void;*/

exports.default = (isOverride /*: IsOverrideType*/, resolveOverride /*: ResolverOverrideType*/) /*: OverrideRequireType*/ => {
  const originalLoad = _module2.default._load;

  // eslint-disable-next-line id-match
  _module2.default._load = function (request /*: string*/, parent /*: ParentType*/) {
    if (isOverride(request, parent)) {
      return resolveOverride(request, parent);
    }

    // eslint-disable-next-line prefer-rest-params
    return originalLoad.apply(this, arguments);
  };

  return () => {
    // eslint-disable-next-line id-match
    _module2.default._load = originalLoad;
  };
};

module.exports = exports['default'];
