// terra-application/lib/workspace/shared/StatusLayout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _reactIntl = require("react-intl");
var _terraDivider = _interopRequireDefault(require("terra-divider"));
var _StatusLayoutModule = _interopRequireDefault(require("./StatusLayout.module.scss"));
var cx = _bind.default.bind(_StatusLayoutModule.default);
var propTypes = {
  /**
   * Sets the glyph and title using a pre-baked variant. One of the following: `no-data`,
   * `no-matching-results`, `not-authorized`, or `error`
   */
  variant: _propTypes.default.oneOf(['no-data', 'no-matching-results', 'not-authorized', 'error']),
  /**
   * The descriptive text, displayed under the title.
   */
  message: _propTypes.default.string,
  /**
   * The actions to render within the StatusLayout.
   */
  children: _propTypes.default.node
};
var StatusLayout = function StatusLayout(_ref) {
  var message = _ref.message,
    variant = _ref.variant,
    children = _ref.children;
  var theme = _react.default.useContext(_terraThemeContext.default);
  var glyphSection;
  if (variant) {
    glyphSection = /*#__PURE__*/_react.default.createElement("div", {
      className: cx('glyph'),
      "data-testid": "status-layout-glyph-".concat(variant)
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: cx(variant, 'glyph-icon')
    }));
  }
  var defaultTitle;
  if (variant) {
    defaultTitle = /*#__PURE__*/_react.default.createElement(_reactIntl.FormattedMessage, {
      id: "terraApplication.statusLayout.".concat(variant)
    });
  }

  // Custom title takes precedence
  var titleSection;
  if (defaultTitle) {
    titleSection = /*#__PURE__*/_react.default.createElement("p", {
      className: cx('title')
    }, defaultTitle);
  }
  var messageSection;
  if (message) {
    messageSection = /*#__PURE__*/_react.default.createElement("p", {
      className: cx('message')
    }, message);
  }
  var actionSection;
  if (children) {
    actionSection = /*#__PURE__*/_react.default.createElement("div", {
      className: cx('actions')
    }, children);
  }
  var dividerSection;
  if (titleSection && (messageSection || actionSection)) {
    dividerSection = /*#__PURE__*/_react.default.createElement("div", {
      className: cx('divider'),
      "data-testid": "status-layout-divider"
    }, /*#__PURE__*/_react.default.createElement(_terraDivider.default, null));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('outer-view', theme.className)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cx('top-space')
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: cx('inner-view')
  }, glyphSection, titleSection, dividerSection, messageSection, actionSection), /*#__PURE__*/_react.default.createElement("div", {
    className: cx('bottom-space')
  }));
};
StatusLayout.propTypes = propTypes;
var _default = StatusLayout;
exports.default = _default;
