// @cerner/terra-dev-site/lib/loader-components/_Badges.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _BadgesModule = _interopRequireDefault(require("./Badges.module.scss"));
var cx = _bind.default.bind(_BadgesModule.default);
var propTypes = {
  /**
   * component name.
   */
  name: _propTypes.default.string.isRequired,
  /**
   * Source code url
   */
  src: _propTypes.default.string,
  /**
   * url to link to
   */
  url: _propTypes.default.string,
  /**
   * The package version.
   */
  version: _propTypes.default.string.isRequired
};
var Badges = function Badges(props) {
  var src = props.src,
    name = props.name,
    url = props.url,
    version = props.version;
  var npmBadge = /*#__PURE__*/_react.default.createElement("a", {
    className: cx('badge'),
    href: url || "https://www.npmjs.org/package/".concat(name, "/v/").concat(version)
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: cx('badge-name')
  }, url ? 'package' : 'npm'), /*#__PURE__*/_react.default.createElement("span", {
    className: cx('badge-version')
  }, "v".concat(version)));
  var srcCodeBadge = src ? /*#__PURE__*/_react.default.createElement("a", {
    className: cx('badge'),
    href: src
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: cx('badge-name')
  }, "github"), /*#__PURE__*/_react.default.createElement("span", {
    className: cx('badge-version')
  }, "source")) : undefined;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cx('badge-container')
  }, npmBadge, srcCodeBadge);
};
Badges.propTypes = propTypes;
var _default = Badges;
exports.default = _default;
