// @cerner/terra-dev-site/lib/loader-components/_PropsTable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RequiredCell = exports.PropNameCell = exports.DescriptionCell = exports.DefaultValueCell = void 0;
Object.defineProperty(exports, "Row", {
  enumerable: true,
  get: function get() {
    return _terraHtmlTable.Row;
  }
});
exports.default = exports.TypeCell = void 0;
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _theme = require("terra-application/lib/theme");
var _terraHtmlTable = _interopRequireWildcard(require("terra-html-table"));
var _PropsTableModule = _interopRequireDefault(require("./PropsTable.module.scss"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* eslint-disable react/forbid-dom-props */

var cx = _bind.default.bind(_PropsTableModule.default);
var propTypes = {
  /**
   * Children for the component.
   */
  children: _propTypes.default.node
};
var requiredCellPropTypes = {
  /**
   * Is the cell required?
   */
  isRequired: _propTypes.default.bool
};
var PropNameCell = function PropNameCell(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Cell, {
    className: cx('bold')
  }, children);
};
exports.PropNameCell = PropNameCell;
PropNameCell.propTypes = propTypes;
var TypeCell = function TypeCell(_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Cell, {
    className: cx('code-block-override')
  }, children);
};
exports.TypeCell = TypeCell;
TypeCell.propTypes = propTypes;
var RequiredCell = function RequiredCell(_ref3) {
  var isRequired = _ref3.isRequired;
  return /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Cell, {
    className: cx([isRequired ? ['required'] : []])
  }, isRequired ? 'required' : 'optional');
};
exports.RequiredCell = RequiredCell;
RequiredCell.propTypes = requiredCellPropTypes;
var DefaultValueCell = function DefaultValueCell(_ref4) {
  var children = _ref4.children;
  return /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Cell, {
    className: cx('code-block-override')
  }, children);
};
exports.DefaultValueCell = DefaultValueCell;
DefaultValueCell.propTypes = propTypes;
var DescriptionCell = function DescriptionCell(_ref5) {
  var children = _ref5.children;
  return /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Cell, null, children);
};
exports.DescriptionCell = DescriptionCell;
DescriptionCell.propTypes = propTypes;
var PropsTable = function PropsTable(_ref6) {
  var children = _ref6.children;
  var theme = (0, _react.useContext)(_theme.ThemeContext);
  return /*#__PURE__*/_react.default.createElement(_terraHtmlTable.default, {
    paddingStyle: "compact",
    className: cx('table', theme.className)
  }, /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Header, {
    className: cx('header')
  }, /*#__PURE__*/_react.default.createElement(_terraHtmlTable.HeaderCell, null, "Prop"), /*#__PURE__*/_react.default.createElement(_terraHtmlTable.HeaderCell, null, "Type"), /*#__PURE__*/_react.default.createElement(_terraHtmlTable.HeaderCell, null, "Required"), /*#__PURE__*/_react.default.createElement(_terraHtmlTable.HeaderCell, null, "Default"), /*#__PURE__*/_react.default.createElement(_terraHtmlTable.HeaderCell, null, "Description")), /*#__PURE__*/_react.default.createElement(_terraHtmlTable.Body, null, children));
};
PropsTable.propTypes = propTypes;
var _default = PropsTable;
exports.default = _default;
