// terra-form-select/lib/native-select/_NativeUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isValuePresent = exports.isCurrentPlaceholder = exports.getOptGroupKeyIndex = exports.getOptGroupKey = exports.getFirstValue = exports.getDisplay = exports.defaultPlaceholderValue = exports.default = void 0;
var defaultPlaceholderValue = '';
exports.defaultPlaceholderValue = defaultPlaceholderValue;
var getDisplay = function getDisplay(findValue, options, isFilterStyle, intl) {
  if (!isFilterStyle && findValue === defaultPlaceholderValue) {
    return intl.formatMessage({
      id: 'Terra.form.select.defaultDisplay'
    });
  }
  if (!options || !options.length) {
    return undefined;
  }
  for (var i = 0; i < options.length; i += 1) {
    var current = options[i];
    if (current.value) {
      if (current.value === findValue) {
        return current.display;
      }
    } else if (current.options) {
      var foundValue = getDisplay(findValue, current.options);
      if (foundValue) {
        return foundValue;
      }
    }
  }
  return undefined;
};
exports.getDisplay = getDisplay;
var getFirstValue = function getFirstValue(options) {
  var isFilterStyle = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  if (!isFilterStyle) {
    return defaultPlaceholderValue;
  }
  if (!options || !options.length) {
    return undefined;
  }
  for (var i = 0; i < options.length; i += 1) {
    var current = options[i];
    if (current.value) {
      return current.value;
    }
    var firstValue = getFirstValue(current.options, isFilterStyle);
    if (firstValue) {
      return firstValue;
    }
  }
  return undefined;
};
exports.getFirstValue = getFirstValue;
var isCurrentPlaceholder = function isCurrentPlaceholder(value) {
  var isFilterStyle = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  return !isFilterStyle && value === defaultPlaceholderValue;
};
exports.isCurrentPlaceholder = isCurrentPlaceholder;
var isValuePresent = function isValuePresent(value) {
  return value !== undefined && value !== null;
};
exports.isValuePresent = isValuePresent;
var getOptGroupKey = function getOptGroupKey(display, index) {
  return "".concat(display, "-").concat(index);
};
exports.getOptGroupKey = getOptGroupKey;
var getOptGroupKeyIndex = function getOptGroupKeyIndex(display, currentKeys) {
  var index = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  if (currentKeys.indexOf(getOptGroupKey(display, index)) >= 0) {
    return getOptGroupKeyIndex(display, currentKeys, index + 1);
  }
  return index;
};
exports.getOptGroupKeyIndex = getOptGroupKeyIndex;
var _default = {
  defaultPlaceholderValue: defaultPlaceholderValue,
  isCurrentPlaceholder: isCurrentPlaceholder,
  isValuePresent: isValuePresent,
  getDisplay: getDisplay,
  getFirstValue: getFirstValue,
  getOptGroupKey: getOptGroupKey,
  getOptGroupKeyIndex: getOptGroupKeyIndex
};
exports.default = _default;
