// terra-form-select/lib/native-select/NativeSelectField.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _terraFormField = _interopRequireDefault(require("terra-form-field"));
var _NativeSelect = _interopRequireDefault(require("./NativeSelect"));
var _NativePropTypes = require("./_NativePropTypes");
var _excluded = ["defaultValue", "disabled", "error", "help", "hideRequired", "isIncomplete", "isInline", "isLabelHidden", "isInvalid", "label", "labelAttrs", "maxWidth", "onChange", "options", "required", "selectAttrs", "selectId", "showOptional", "value"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var propTypes = {
  /**
   * The default value of the select. Can be a string, or number.
   */
  defaultValue: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  /**
   * Whether the input is disabled.
   */
  disabled: _propTypes.default.bool,
  /**
   * Error message displayed when the select is invalid.
   */
  error: _propTypes.default.node,
  /**
   * Help message to display with the select.
   */
  help: _propTypes.default.node,
  /**
   * Whether to hide the required indicator on the label.
   */
  hideRequired: _propTypes.default.bool,
  /**
   * Whether the field displays as Incomplete. Use when no value has been provided. _(usage note: `required` must also be set)_.
   */
  isIncomplete: _propTypes.default.bool,
  /**
   * Whether the field is displayed inline. Displays block by default.
   */
  isInline: _propTypes.default.bool,
  /**
   * Whether the label is hidden. Allows hiding the label while meeting accessibility guidelines.
   */
  isLabelHidden: _propTypes.default.bool,
  /**
   * Whether the field displays as Invalid. Use when value does not meet validation pattern.
   */
  isInvalid: _propTypes.default.bool,
  /**
   * The field label.
   */
  label: _propTypes.default.node.isRequired,
  /**
   * Additional attributes to spread onto the label.
   */
  // eslint-disable-next-line react/forbid-prop-types
  labelAttrs: _propTypes.default.object,
  /**
   * Set the max-width of a field using `length` or `%`.  Best practice recommendation to never exceed
   * a rendered value of 1020px. _(Note: Providing custom inline styles will take precedence.)_
   */
  maxWidth: _propTypes.default.string,
  /**
   * Callback function triggered when the select value changes. function(event)
   */
  onChange: _propTypes.default.func,
  /**
   * The array of select options and opt groups.
   */
  options: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_NativePropTypes.optionPropType, _NativePropTypes.optGroupPropType])),
  /**
   * Whether the field is required.
   */
  required: _propTypes.default.bool,
  /**
   * Additional attributes to spread onto the select.
   */
  // eslint-disable-next-line react/forbid-prop-types
  selectAttrs: _propTypes.default.object,
  /**
   * The Select identifier. Links the htmlFor of the field to the select identifier.
   */
  selectId: _propTypes.default.string.isRequired,
  /**
   * Whether to append the 'optional' label to a non-required field label.
   */
  showOptional: _propTypes.default.bool,
  /**
   * The value of the select. Can be a string or number.
   */
  value: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
};
var defaultProps = {
  disabled: false,
  hideRequired: false,
  isIncomplete: false,
  isInline: false,
  isInvalid: false,
  isLabelHidden: false,
  required: false,
  showOptional: false
};
var NativeSelectField = function NativeSelectField(_ref) {
  var defaultValue = _ref.defaultValue,
    disabled = _ref.disabled,
    error = _ref.error,
    help = _ref.help,
    hideRequired = _ref.hideRequired,
    isIncomplete = _ref.isIncomplete,
    isInline = _ref.isInline,
    isLabelHidden = _ref.isLabelHidden,
    isInvalid = _ref.isInvalid,
    label = _ref.label,
    labelAttrs = _ref.labelAttrs,
    maxWidth = _ref.maxWidth,
    onChange = _ref.onChange,
    options = _ref.options,
    required = _ref.required,
    selectAttrs = _ref.selectAttrs,
    selectId = _ref.selectId,
    showOptional = _ref.showOptional,
    value = _ref.value,
    customProps = _objectWithoutProperties(_ref, _excluded);
  var helpText = help ? /*#__PURE__*/_react.default.createElement("span", null, help) : undefined;
  var ariaDescriptionIds;
  if (help && error && isInvalid) {
    ariaDescriptionIds = "".concat(selectId, "-error ").concat(selectId, "-help");
  } else {
    if (help) {
      ariaDescriptionIds = "".concat(selectId, "-help");
    }
    if (error && isInvalid) {
      ariaDescriptionIds = "".concat(selectId, "-error");
    }
  }
  return /*#__PURE__*/_react.default.createElement(_terraFormField.default, _extends({}, customProps, {
    label: label,
    labelAttrs: labelAttrs,
    error: error,
    help: helpText,
    hideRequired: hideRequired,
    required: required,
    showOptional: showOptional,
    isInvalid: isInvalid,
    isInline: isInline,
    isLabelHidden: isLabelHidden,
    htmlFor: selectId,
    maxWidth: maxWidth
  }), /*#__PURE__*/_react.default.createElement(_NativeSelect.default, {
    attrs: selectAttrs,
    id: selectId,
    ariaDescribedBy: ariaDescriptionIds,
    ariaLabel: label,
    disabled: disabled,
    isIncomplete: isIncomplete,
    isInvalid: isInvalid,
    defaultValue: defaultValue,
    required: required,
    onChange: onChange,
    options: options,
    value: value
  }));
};
NativeSelectField.propTypes = propTypes;
NativeSelectField.defaultProps = defaultProps;
var _default = NativeSelectField;
exports.default = _default;
