// terra-html-table/lib/Table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _classnames = _interopRequireDefault(require("classnames"));
var _bind = _interopRequireDefault(require("classnames/bind"));
var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));
var _TableModule = _interopRequireDefault(require("./Table.module.scss"));
var _excluded = ["children", "disableStripes", "paddingStyle"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var cx = _bind.default.bind(_TableModule.default);
var propTypes = {
  /**
   * The child content for the table consisting of either a TableHeader and/or a TableBody.
   */
  children: _propTypes.default.node.isRequired,
  /**
   * Whether or not the rows should be zebra striped.
   */
  disableStripes: _propTypes.default.bool,
  /**
   * Indicates the desired divider styles to apply to a row and its children.
   * One of `'none'`, `'standard'`, `'compact'`.
   */
  paddingStyle: _propTypes.default.oneOf(['none', 'standard', 'compact'])
};
var defaultProps = {
  disableStripes: false,
  paddingStyle: 'none'
};
var Table = function Table(_ref) {
  var children = _ref.children,
    disableStripes = _ref.disableStripes,
    paddingStyle = _ref.paddingStyle,
    customProps = _objectWithoutProperties(_ref, _excluded);
  var theme = _react.default.useContext(_terraThemeContext.default);
  var tableClassNames = (0, _classnames.default)(cx('table', {
    striped: !disableStripes
  }, {
    'padding-standard': paddingStyle === 'standard'
  }, {
    'padding-compact': paddingStyle === 'compact'
  }, theme.className), customProps.className);
  return /*#__PURE__*/_react.default.createElement("table", _extends({}, customProps, {
    className: tableClassNames
  }), children);
};
Table.propTypes = propTypes;
Table.defaultProps = defaultProps;
var _default = Table;
exports.default = _default;
