// terra-application/lib/application-navigation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ApplicationNavigationActionsContext", {
  enumerable: true,
  get: function get() {
    return _ApplicationNavigationActionsContext.default;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "extensionItemsPropType", {
  enumerable: true,
  get: function get() {
    return _propTypes.extensionItemsPropType;
  }
});
Object.defineProperty(exports, "navigationItemsPropType", {
  enumerable: true,
  get: function get() {
    return _propTypes.navigationItemsPropType;
  }
});
Object.defineProperty(exports, "titleConfigPropType", {
  enumerable: true,
  get: function get() {
    return _propTypes.titleConfigPropType;
  }
});
Object.defineProperty(exports, "userConfigPropType", {
  enumerable: true,
  get: function get() {
    return _propTypes.userConfigPropType;
  }
});
Object.defineProperty(exports, "utilityItemsPropType", {
  enumerable: true,
  get: function get() {
    return _propTypes.utilityItemsPropType;
  }
});
var _propTypes = require("./private/utils/propTypes");
var _ApplicationNavigationActionsContext = _interopRequireDefault(require("./ApplicationNavigationActionsContext"));
var _ApplicationNavigation = _interopRequireDefault(require("./ApplicationNavigation"));
var _default = _ApplicationNavigation.default;
exports.default = _default;
