// terra-form-select/lib/MultiSelect.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _reactIntl = require("react-intl");
var _Frame = _interopRequireDefault(require("./multiple/Frame"));
var _Option = _interopRequireDefault(require("./shared/_Option"));
var _OptGroup = _interopRequireDefault(require("./shared/_OptGroup"));
var _Tag = _interopRequireDefault(require("./shared/_Tag"));
var _SelectUtil = _interopRequireDefault(require("./shared/_SelectUtil"));
var _excluded = ["children", "defaultValue", "onChange", "placeholder", "required", "value", "intl", "inputId"];
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
var propTypes = {
  /**
   * The dropdown menu options.
   */
  children: _propTypes.default.node,
  /**
   * The default selected value. Can be a string, number, or array of strings/numbers.
   */
  defaultValue: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.array]),
  /**
   * Whether the select is disabled.
   */
  disabled: _propTypes.default.bool,
  /**
   * Additional attributes to spread onto the dropdown. ( Style, ClassNames, etc.. )
   */
  // eslint-disable-next-line react/forbid-prop-types
  dropdownAttrs: _propTypes.default.object,
  /**
   * @private
   * The intl object containing translations. This is retrieved from the context automatically by injectIntl.
   */
  intl: _propTypes.default.shape({
    formatMessage: _propTypes.default.func
  }).isRequired,
  /**
   * Whether the select displays as Incomplete. Use when no value has been provided. _(usage note: `required` must also be set)_.
   */
  isIncomplete: _propTypes.default.bool,
  /**
   * Whether the select displays as Invalid. Use when value does not meet validation pattern.
   */
  isInvalid: _propTypes.default.bool,
  /**
   * Ensures touch accessibility by rendering the dropdown inline without a portal.
   *
   * Note: When enabled the dropdown will clip if rendered within a container that has an overflow: hidden ancestor.
   * The dropdown may also appear beneath content if rendered within a container that has an overflow: auto ancestor.
   */
  isTouchAccessible: _propTypes.default.bool,
  /**
   * The max height of the dropdown.
   */
  maxHeight: _propTypes.default.number,
  /**
   * @private The maximum number of options that can be selected. A value less than 2 will be ignored.
   */
  maxSelectionCount: _propTypes.default.number,
  /**
   * Content to display when no results are found.
   */
  noResultContent: _propTypes.default.node,
  /**
   * Callback function triggered when the select loses focus. function(event)
   */
  onBlur: _propTypes.default.func,
  /**
   * Callback function triggered when the value changes. function(value)
   */
  onChange: _propTypes.default.func,
  /**
   * Callback function triggered when the select is clicked. function(event)
   */
  onClick: _propTypes.default.func,
  /**
   * Callback function triggered when an option is deselected. function(value)
   */
  onDeselect: _propTypes.default.func,
  /**
   * Callback function triggered when the select receives focus. function(event)
   */
  onFocus: _propTypes.default.func,
  /**
   * Callback function triggered when the search criteria changes. function(searchValue)
   */
  onSearch: _propTypes.default.func,
  /**
   * Callback function triggered when an option is selected. function(value)
   */
  onSelect: _propTypes.default.func,
  /**
   * Callback function invoked for each option on search change. function(searchValue, option)
   */
  optionFilter: _propTypes.default.func,
  /**
   * Placeholder text.
   */
  placeholder: _propTypes.default.string,
  /**
   * Whether the field is required.
   */
  required: _propTypes.default.bool,
  /**
   * The selected value. Can be a string, number, or array of strings/numbers.
   */
  value: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.array]),
  /**
   * @private
   * The id of the input field.
   */
  inputId: _propTypes.default.string
};
var defaultProps = {
  children: undefined,
  defaultValue: undefined,
  disabled: false,
  dropdownAttrs: undefined,
  isIncomplete: false,
  isInvalid: false,
  isTouchAccessible: false,
  maxSelectionCount: undefined,
  noResultContent: undefined,
  onChange: undefined,
  onDeselect: undefined,
  onSearch: undefined,
  onSelect: undefined,
  optionFilter: undefined,
  placeholder: undefined,
  required: false,
  value: undefined,
  inputId: undefined
};
var MultiSelect = /*#__PURE__*/function (_React$Component) {
  _inherits(MultiSelect, _React$Component);
  var _super = _createSuper(MultiSelect);
  function MultiSelect(props) {
    var _this;
    _classCallCheck(this, MultiSelect);
    _this = _super.call(this, props);
    var value = props.value,
      defaultValue = props.defaultValue;
    _this.state = {
      value: _SelectUtil.default.defaultValue({
        defaultValue: defaultValue,
        value: value,
        multiple: true
      })
    };
    _this.display = _this.display.bind(_assertThisInitialized(_this));
    _this.handleChange = _this.handleChange.bind(_assertThisInitialized(_this));
    _this.handleDeselect = _this.handleDeselect.bind(_assertThisInitialized(_this));
    _this.handleSelect = _this.handleSelect.bind(_assertThisInitialized(_this));
    return _this;
  }

  /**
   * Communicates changes to the value.
   * @param {array|number|string} value - The value resulting from a change.
   */
  _createClass(MultiSelect, [{
    key: "handleChange",
    value: function handleChange(value) {
      if (this.props.value === undefined) {
        this.setState({
          value: value
        });
      }
      if (this.props.onChange) {
        this.props.onChange(value);
      }
    }

    /**
     * Communicates the removal of a value from the selected options.
     * @param {number|string} value - The value to be removed.
     */
  }, {
    key: "handleDeselect",
    value: function handleDeselect(value) {
      this.handleChange(_SelectUtil.default.deselect(this.props, this.state, value));
      if (this.props.onDeselect) {
        this.props.onDeselect(value);
      }
    }

    /**
     * Communicates the selection of a value.
     * @param {number|string} value - The value of the selected option.
     * @param {ReactNode} option - The selected option.
     */
  }, {
    key: "handleSelect",
    value: function handleSelect(value, option) {
      var newValue = [].concat(_toConsumableArray(_SelectUtil.default.value(this.props, this.state)), [value]);
      this.handleChange(newValue);
      if (this.props.onSelect) {
        this.props.onSelect(value, option);
      }
    }

    /**
     * Returns the appropriate variant display
     */
  }, {
    key: "display",
    value: function display() {
      var _this2 = this;
      var selectValue = _SelectUtil.default.value(this.props, this.state);
      return selectValue.map(function (tag) {
        return /*#__PURE__*/_react.default.createElement(_Tag.default, {
          value: tag,
          key: tag,
          onDeselect: _this2.handleDeselect
        }, _SelectUtil.default.valueDisplay(_this2.props, tag));
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        children = _this$props.children,
        defaultValue = _this$props.defaultValue,
        onChange = _this$props.onChange,
        placeholder = _this$props.placeholder,
        required = _this$props.required,
        value = _this$props.value,
        intl = _this$props.intl,
        inputId = _this$props.inputId,
        otherProps = _objectWithoutProperties(_this$props, _excluded);
      var defaultPlaceholder = intl.formatMessage({
        id: 'Terra.form.select.defaultDisplay'
      });
      var selectPlaceholder = placeholder === undefined ? defaultPlaceholder : placeholder;
      return /*#__PURE__*/_react.default.createElement(_Frame.default, _extends({}, otherProps, {
        "data-terra-select": true,
        value: _SelectUtil.default.value(this.props, this.state),
        display: this.display(),
        onDeselect: this.handleDeselect,
        onSelect: this.handleSelect,
        placeholder: selectPlaceholder,
        required: required,
        totalOptions: _SelectUtil.default.getTotalNumberOfOptions(children),
        inputId: inputId
      }), children);
    }
  }]);
  return MultiSelect;
}(_react.default.Component);
MultiSelect.Option = _Option.default;
MultiSelect.OptGroup = _OptGroup.default;
MultiSelect.propTypes = propTypes;
MultiSelect.defaultProps = defaultProps;
MultiSelect.isSelect = true;
var _default = (0, _reactIntl.injectIntl)(MultiSelect);
exports.default = _default;
