// terra-application/lib/terra-dev-site/test/application-navigation/private/application-navigation/Title.test.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _ApplicationNavigation = _interopRequireDefault(require("../../../../../application-navigation/private/ApplicationNavigation"));
var _FixedHeight = _interopRequireDefault(require("../common/FixedHeight"));
var _Logo = _interopRequireDefault(require("../common/Logo"));
var titleConfig = {
  title: 'Test Title'
};
var titleConfigHeadLine = {
  title: 'Test Title',
  headline: 'Test HeadLine'
};
var titleConfigSubline = {
  title: 'Test Title',
  subline: 'Test Subline'
};
var titleConfigElement = {
  title: 'Test Title',
  element: /*#__PURE__*/_react.default.createElement(_Logo.default, null)
};
var titleConfigHideTitle = {
  title: 'Test Title',
  hideTitleWithinDrawerMenu: true
};

// TODO: remove terra-application after it is incorporated into dev-site for themes or themes are co-located
var TitleTest = function TitleTest() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_FixedHeight.default, null, /*#__PURE__*/_react.default.createElement(_ApplicationNavigation.default, {
    id: "test-default",
    titleConfig: titleConfig
  })), /*#__PURE__*/_react.default.createElement(_FixedHeight.default, null, /*#__PURE__*/_react.default.createElement(_ApplicationNavigation.default, {
    id: "test-headline",
    titleConfig: titleConfigHeadLine
  })), /*#__PURE__*/_react.default.createElement(_FixedHeight.default, null, /*#__PURE__*/_react.default.createElement(_ApplicationNavigation.default, {
    id: "test-subline",
    titleConfig: titleConfigSubline
  })), /*#__PURE__*/_react.default.createElement(_FixedHeight.default, null, /*#__PURE__*/_react.default.createElement(_ApplicationNavigation.default, {
    id: "test-element",
    titleConfig: titleConfigElement
  })), /*#__PURE__*/_react.default.createElement(_FixedHeight.default, null, /*#__PURE__*/_react.default.createElement(_ApplicationNavigation.default, {
    id: "test-title",
    titleConfig: titleConfigHideTitle
  })));
};
var _default = TitleTest;
exports.default = _default;
