// terra-application/lib/terra-dev-site/app/components.2/NotificationBannerPropsTable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
/* eslint-disable react/no-unused-prop-types */
var propTypes = {
  /**
   * The text and corresponding callback to populate the action button of the banner.
   */
  bannerAction: _propTypes.default.shape({
    /**
     * The text to display in the banner button.
     */
    text: _propTypes.default.string,
    /**
     * The Callback function triggered when the action button is clicked. No parameters are passed.
     */
    onClick: _propTypes.default.func
  }),
  /**
   * The message content to display in the banner.
   */
  description: _propTypes.default.node,
  /**
   * Callback function triggered when the dismiss button is clicked. The presence of this prop will cause
   * the dismiss button to be included on the banner. No parameters are passed.
   */
  onRequestClose: _propTypes.default.func,
  /**
   * The variant of notification banner to be rendered. This renders the banner with the corresponding header and icon to the
   * variant concept.
   */
  variant: _propTypes.default.oneOf(['hazard-high', 'hazard-medium', 'hazard-low', 'error', 'unsatisfied', 'unverified', 'custom']).isRequired,
  /**
   * The pieces to populate a banner when `variant="custom"`.
   */
  custom: _propTypes.default.shape({
    /**
     * The keyword used to represent & emphasis the intention of banner description that is being shown to the user.
     */
    signalWord: _propTypes.default.string,
    /**
     * The class name used to set the icon as the background image to be used as the icon in the banner.
     */
    iconClassName: _propTypes.default.string
  })
};
/* eslint-enable react/no-unused-prop-types */

var PropTypesExample = function PropTypesExample() {
  return /*#__PURE__*/_react.default.createElement("div", null);
};
PropTypesExample.propTypes = propTypes;
var _default = PropTypesExample;
exports.default = _default;
